pg_dphyp (#6)
ashenBlade/habr-posts: pg_dphyp/post.md

# pg_dphyp

Приветствую!

Как разработчик баз данных я увлекаюсь базами данных и, читая [красную книгу](https://www.redbook.io/), что-то во мне щелкнуло и захотел глубже поизучать планировщик. Спустя некоторое время сидел над алгоритмами соединений таблиц и наткнулся на алгоритм DPhyp. Его используют многие современные (и не очень) базы данных. Мне стало интересно: а есть ли в PostgreSQL такое? К великому удивлению, нет. Ну, если чего-то нет, то это надо создать самому. Закатал рукава, создал репозиторий и принялся писать код.

Эта статья не о DPhyp и том, как он работает, а о том, с чем мне пришлось столкнуться в процессе написания соответствующего расширения для PostgreSQL. Но вначале немного теории.

## Планировщики запросов

Планировщик в базах данных это, пожалуй, самый сложный и важный компонент системы, особенно, если мы говорим о терабайтах данных: неважно насколько быстрое железо стоит в серверной - если планировщик чуть-чуть ошибся и использует последовательное сканирование вместо индекса, то приходите за результатом через неделю. И в этом сложном компоненте можно выделить ядро - перебор JOIN'ов.

> Это важная часть, так как по большей часть правильный выбор порядка соединения таблиц оказывает наибольшее влияние на стоимость всего запроса. Например, вот такой запрос:

Reviewer: ```suggestion
Это важная часть, так как правильный выбор порядка соединения таблиц оказывает наибольшее влияние на стоимость всего запроса. Например, вот такой запрос:
```

```sql
SELECT *
FROM t1 
JOIN t2 ON t1.x = t2.x
JOIN t3 ON t2.x = t3.x
JOIN t4 ON t3.x = t4.x;
```

> Имеет 14 возможных комбинаций соединений таблиц. А в общем случае - это число возможных представлений бинарного дерева из `N` узлов, где под узлами понимаются таблицы, [число Каталана](https://ru.wikipedia.org/wiki/Числа_Каталана), которое уже для 7 таблиц равно 429, а дальше для 8 - 1430. Стоит ли говорить, что с определенного момента оно становится насколько огромным, что практически невозможно найти оптимальный план простым перебором всех возможных комбинаций.

Reviewer: ```suggestion
Имеет 14 возможных комбинаций соединений таблиц. А в общем случае - это число возможных представлений бинарного дерева из `N` узлов, где под узлами понимаются таблицы, [число Каталана](https://ru.wikipedia.org/wiki/Числа_Каталана), которое уже для 7 таблиц равно 429, а дальше для 8 - 1430. Стоит ли говорить, что с определенного момента оно становится настолько огромным, что практически невозможно найти оптимальный план простым перебором всех возможных комбинаций.
```

То, как мы собираемся выполнять перебор и определяет ~~наш лагерь~~ архитектуру планировщика: top-down или bottom-up.

> Top-down подход - это подход сверху вниз (еще называют goal-oriented, ориентированный на цель). Мы заходим в корень, понимаем его "цель" (есть ли группировка/фильтрация, что на выходе и т.д.) и на основании этого создают план запроса. Преимуществом такого подхода является то, что у нас на руках полный контекст и мы можем его использовать. Примером может служить планировщик (грубо говоря, это архитектура) cascades, который используется в MS SQL Server.

Reviewer: ```suggestion
Top-down — это подход сверху вниз (еще называют goal-oriented, ориентированный на цель). Мы заходим в корень, понимаем его "цель" (есть ли группировка/фильтрация, что на выходе и т.д.) и на основании этого создают план запроса. Преимуществом такого подхода является то, что у нас на руках полный контекст и мы можем его использовать. Примером может служить планировщик (грубо говоря, это архитектура) cascades, который используется в Microsoft SQL Server.
```

> Bottom-up - это противоположный лагерь, где вначале планируют все JOIN'ы, а только потом навешивают сортировку/группировку и другие операторы. Этот подход используются во многих базах данных, например, в том же самом PostgreSQL. Преимуществом этого подхода является масштабируемость, так как он позволяет планировать огромное количество JOIN'ов, например, в статье [Adaptive Optimization of Very Large Join Queries](https://db.in.tum.de/~radke/papers/hugejoins.pdf) представляется подход, который комбинируя несколько разных алгоритмов позволяет выполнять планирование запросов с несколькими тысячами таблиц. Примером кто так делает в статье указывается SAP, который из-за постоянного использования представлений внутри других представлений может создать запрос использующий тысячи обычных таблиц.

Reviewer: ```suggestion
Bottom-up — это противоположный подход, где вначале планируют все JOIN'ы, а только потом навешивают сортировку/группировку и другие операторы. Этот подход используются во многих базах данных, например, в том же PostgreSQL. Преимуществом этого подхода является масштабируемость, так как он позволяет планировать огромное количество JOIN'ов, например, в статье [Adaptive Optimization of Very Large Join Queries](https://db.in.tum.de/~radke/papers/hugejoins.pdf) представляется подход, который комбинируя несколько разных алгоритмов позволяет выполнять планирование запросов с несколькими тысячами таблиц. Примером того, кто так делает в статье приводится SAP, который из-за постоянного использования представлений внутри других представлений может создать запрос использующий тысячи обычных таблиц.
```

> Сейчас мы акцентируем внимание на последнем подходе - bottom-up, а именно на используемых алгоритмах. Но этих алгоритмов довольно много, поэтому вначале рассмотри алгоритмы, используемые в PostgreSQL.

Reviewer: ```suggestion
Сейчас мы акцентируем внимание на втором подходе — bottom-up, а именно на используемых им алгоритмах. Но этих алгоритмов довольно много, поэтому вначале рассмотрим только те алгоритмы, которые используются в PostgreSQL.
```

Его идея довольно проста - чтобы создать JOIN из `i` таблиц отношений нужно соединить другие отношения, которые в сумме количества таблиц дадут этот `i`. Например, для `4` нам нужно соединить: `1` и `3`, `2` и `2`. Собственно, это и есть динамическое программирование - ответ текущего шага зависит от ответа прошлых, ну а базой выступают отношения размером `1`, то есть обычные таблицы. Этот алгоритм неплохо себя показывает на OLTP нагрузке (когда таблиц мало) и дает практически оптимальные планы запросов, но проблемы начинаются когда таблиц становится слишком много.

Как можете заметить, сложность этого алгоритма экспоненциальная, так как на каждом следующем шаге нам предстоит обработать еще больше пар отношений. Разные базы данных борются с этим по разному, но в PostgreSQL пошли путем использования другого алгоритма.

### GEQO

> GEQO, Genetic Query Optimizer - это генетический алгоритм поиска оптимального плана запроса. Если вы попробуете запустить в PostgreSQL запрос сначала на 12 таблиц, а после на 13, то удивитесь, что затраченное время снизилось с нескольких секунд, до почти *десяти миллисекунд*. Почему?

Reviewer: ```suggestion
GEQO (Genetic Query Optimizer) — это генетический алгоритм поиска оптимального плана запроса. Если вы попробуете запустить в PostgreSQL запрос сначала на 12 таблиц, а после на 13, то удивитесь, что затраченное время снизилось с нескольких секунд, до почти *десяти миллисекунд*. Почему?
```

Ответ прост - это рандомизированный алгоритм. Его работу можно описать просто: вначале строим *хоть какой-нибудь* план запроса, а затем проводим несколько итераций (определяется конфигурацией), в каждой из которых случайно меняем какие-нибудь узлы, и в следующую итерацию идет план с лучшей стоимостью. Отсюда и название - генетический, так как в следующее поколение (итерацию) переходят сильнейшие (самые дешевые).

## DPhyp

Переходим к основной теме - DPhyp.

> DPhyp - это алгоритм динамического программирования для перебора JOIN'ов. Его основная идея заключается в том, что сам запрос содержит указания того, как должно происходить соединение таблиц. Так почему-бы его не использовать? Основная проблема в самом представлении запроса. Ранее, я оговорился, что запрос можно представить в виде графа, но можно ли так сделать для целей перебора JOIN'ов? Чтобы понять в чем трудность посмотрим на пример из статьи:

Reviewer: ```suggestion
DPhyp - это алгоритм динамического программирования для перебора JOIN'ов. Его основная идея заключается в том, что сам запрос содержит указания того, как должно происходить соединение таблиц. Так почему бы его не использовать? Основная проблема в самом представлении запроса. Ранее, я оговорился, что запрос можно представить в виде графа, но можно ли так сделать для целей перебора JOIN'ов? Чтобы понять в чем трудность посмотрим на пример из статьи:
```

```sql
SELECT *
FROM R1, R2, R3, R4, R5, R6
WHERE R1.x = R2.x AND R2.x = R3.x AND R4.x = R5.x AND R5.x = R6.x AND
      R1.x + R2.x + R3.x = R4.x + R5.x + R6.x
```

Да, мы видим, что есть несколько *явно* соединенных между собой таблиц - для них мы можем создать ребра в нашем графе (например, `R1 - R2`), но что делать с последним предикатом, который по факту соединяет множества таблиц?

> Эту проблему и решает DPhyp - алгоритм динамического программирования (DP), основанных на *гиперграфах* (hyp - hypergraph).

Reviewer: ```suggestion
Эту проблему и решает DPhyp - алгоритм динамического программирования (DP), основанный на *гиперграфах* (hyp - hypergraph).
```

Пугаться не стоит, все довольно просто. Для начала, предполагаю, что вы знакомы с обычным графом - множество узлов, соединенных между собой ребрами. Гиперграф - это множество гиперузлов, соединенных между собой гиперребрами:

- гиперузел (hypernode) - это *множество* обычных узлов

> - гиперребро (hyperedge) - это ребро, *соединяющее 2 гиперузла*

Reviewer: Числительные до десяти лучше записывать буквами, после — цифрами.
```suggestion
- гиперребро (hyperedge) - это ребро, *соединяющее два гиперузла*
```

В запросе из примера у нас имеются следующие гиперребра:

1. `{R1} - {R2}`

> 2. `{R2} - {R3}`

Reviewer: Если это маркдаун, то все элементы нумерованного списка пишутся только через `1.`. При рендере они сами верно пронумеруются.

3. `{R4} - {R5}`
4. `{R5} - {R6}`
5. `{R1, R2, R3} - {R4, R5, R6}`

> Если в гиперузле только 1 отношение, то его называют *простым гиперузлом*. Аналогично, если гиперребро соединяет 2 простых гиперузла, то это *простой гиперузел*. Первые 4 - это простые гиперребра.

Reviewer: Аналогично, числительные до десяти лучше записывать буквами, после — цифрами.
```suggestion
Если в гиперузле только одно отношение, то его называют *простым гиперузлом*. Аналогично, если гиперребро соединяет два простых гиперузла, то это *простой гиперузел*. Первые четыре - это простые гиперребра.
```

> В итоге, чтобы создать план для множества из `i` узлов нам нужно обойти уже не все возможные пары, дающие `i` в сумме, а все пары гиперузлов, которые в объединении дают тоже множество. Такая пара - это пара из 2-ух непересекающихся множеств: `connected subgraph` (csg, подграф) и `connected complement` (cmp, дополнение). Эти аббревиатуры вы увидите еще не раз.

Reviewer: ```suggestion
В итоге, чтобы создать план для множества из `i` узлов, нам нужно обойти уже не все возможные пары, дающие `i` в сумме, а все пары гиперузлов, которые в объединении дают то же множество. Такая пара — это пара из двух непересекающихся множеств: `connected subgraph` (csg, подграф) и `connected complement` (cmp, дополнение). Эти аббревиатуры вы увидите еще не раз.
```

> Но не все так просто и, чтобы это заработало оптимально, требуется небольшое ограничение - порядок узлов. Над узлами (т.е. таблицами) должно быть отношение порядка, грубо говоря, они должны быть пронумерованы (что используется чаще всего) для того, чтобы потом мы могли их сравнить и отсортировать. Чтобы понять зачем рассмотрим ядро алгоритма - соседство.

Reviewer: ```suggestion
Но не все так просто и, чтобы это заработало оптимально, требуется небольшое ограничение — порядок узлов. Над узлами (т. е. таблицами) должно быть отношение порядка, грубо говоря, они должны быть пронумерованы (что используется чаще всего) для того, чтобы потом мы могли их сравнить и отсортировать. Чтобы понять зачем, рассмотрим ядро алгоритма — соседство.
```

> В процессе работы алгоритма, чтобы переходить от одного гиперузла к другому мы используем соседей (neighborhood). В статье дается и математическое определение, но проще всего сказать, что соседи для какого-либо гиперузла - это множество других достижимых узлов. Также требуется, чтобы это множество было минимальным, в противном случае мы будем обрабатывать одни и те же узлы несколько раз. Вот тут и нужен порядок - когда обходим ребра для нахождения соседей, то в множество добавляем только *представителя* гиперузла, его *минимальный элемент*. Дальше нам нужно просто проверить, что другие ребра не содержат уже добавленные узлы.

Reviewer: ```suggestion
В процессе работы алгоритма, чтобы переходить от одного гиперузла к другому, мы используем соседей (neighborhood). В статье дается и математическое определение, но проще всего сказать, что соседи для какого-либо гиперузла — это множество других достижимых узлов. Также требуется, чтобы это множество было минимальным, в противном случае мы будем обрабатывать одни и те же узлы несколько раз. Вот тут и нужен порядок — когда обходим ребра для нахождения соседей, то в множество добавляем только *представителя* гиперузла, его *минимальный элемент*. Дальше нам нужно просто проверить, что другие ребра не содержат уже добавленные узлы.
```

И последняя важная деталь алгоритма - исключенное множество (excluded set). В DPsize в качестве оптимизации итерирование по дополняющим парам мы начинаем не с 0, а со следующего индекса, или в противном случае попытаемся соединить себя с собой же, а некоторые пары обработать дважды. Здесь примерно та же идея - мы ведем учет узлов, которые не стоит рассматривать (исключенные) и проверяем это практически везде (даже при нахождении соседей). Благодаря этому, мы не рассматривать одно и то же множество дважды.

> Основная логика алгоритма в статье представлена в виде 4 функций, но вкратце ядро алгоритма можно описать так: нам нужно найти csg-cmp пару, которая в объединении дает весь запрос, поэтому с помощью поиска соседей будет поочередно/рекурсивно увеличивать csg и cmp. Вопрос только в том с чего начать. На это ответ тоже простой - мы начинаем итерироваться по всем узлам начиная *с конца*, а при дальнейшем вызове в *excluded* записываем все узлы, которые меньше текущего. В итоге, мы начинаем с простого гиперузла, который никто не рассматривал и затем рекурсивно его расширяем, находим cmp для него с помощью соседей.

Reviewer: ```suggestion
Основная логика алгоритма в статье представлена в виде четырех функций, но вкратце ядро алгоритма можно описать так: нам нужно найти csg-cmp пару, которая в объединении дает весь запрос, поэтому с помощью поиска соседей будет поочередно/рекурсивно увеличивать csg и cmp. Вопрос только в том, с чего начать. На это ответ тоже простой — мы начинаем итерироваться по всем узлам начиная *с конца*, а при дальнейшем вызове в *excluded* записываем все узлы, которые меньше текущего. В итоге, мы начинаем с простого гиперузла, который никто не рассматривал, а затем рекурсивно его расширяем и находим cmp для него с помощью соседей.
```

Собственно, теперь все готово, чтобы понять функции:

> - `Solve` - входная точка алгоритма, просто итерируемся по простым гиперузлам с конца и вызываем `EmitCsg` и `EnumerateCsgRecursive`

Reviewer: ```suggestion
- `Solve` — входная точка алгоритма. Просто итерируемся по простым гиперузлам с конца и вызываем `EmitCsg` и `EnumerateCsgRecursive`.
```

> - `EmitCsg` - принимает уже *фиксированный* csg, для которого находит подходящий cmp, а затем вызывает `EmitCsgCmp` и/или `EnumerateCmpRecursive`

Reviewer: ```suggestion
- `EmitCsg` — принимает уже *фиксированный* csg, для которого находит подходящий cmp, а затем вызывает `EmitCsgCmp` и/или `EnumerateCmpRecursive`.
```

> - `EmitCsgRecursive` - принимает csg, который расширяет с помощью соседей, затем вызывает `EmitCsg` и/или `EnumerateCsgRecursive`

Reviewer: ```suggestion
- `EmitCsgRecursive` — принимает csg, который расширяет с помощью соседей, затем вызывает `EmitCsg` и/или `EnumerateCsgRecursive`.
```

> - `EnumerateCmpRecursive` - принимает *фиксированный* csg и cmp, с помощью соседей расширяет cmp, затем вызывает `EmitCsgCmp` и/или `EnumerateCmpRecursive`

Reviewer: ```suggestion
- `EnumerateCmpRecursive` — принимает *фиксированный* csg и cmp, с помощью соседей расширяет cmp, затем вызывает `EmitCsgCmp` и/или `EnumerateCmpRecursive`.
```

> - `EmitCsgCmp` - создает готовый план для *фиксированных* csg и cmp

Reviewer: ```suggestion
- `EmitCsgCmp` — создает готовый план для *фиксированных* csg и cmp.
```

В общем, это все. Основная идея алгоритма довольно проста и понятна. В качестве примера в статье также есть пошаговая иллюстрация работы алгоритма для запроса из примера:

![Трейсинг алгоритма для запроса из примера](./img/paper-trace.png)

Что происходит:

1. Итерирование происходит в обратном порядке, поэтому начинаем с отношения `R6` (который самый большой по индексу), но для него соседей нет, так как все остальные узлы находятся в исключенном множестве.

> 2. Переходим к `R5`

Reviewer: Поставь точки в конце каждого пункта списка.
И если это маркдаун, то все элементы нумерованного списка пишутся только через `1.`. При рендере они сами верно пронумеруются.

3. Для него нашли соседа `R6` и создаем из него CMP `{R6}`
4. Теперь увеличиваем сам CSG до `{R5, R6}`
5. Переходим к `R4`
6. Для него есть сосед `{R5}`, для которого создаем CMP `{R5}`
7. После увеличиваем этот CMP до `{R5, R6}`, так как `R6` сосед `R5`

> 8. Возвращаемся к 6 итерации и вместо увеличиваем сам CSG до `{R4, R5}`

Reviewer: "к шестой итерации"
И не написано вместо чего мы "увеличиваем сам CSG до `{R4, R5}`"

9. Для уже этого подграфа мы нашли CMP `{R6}`
10. Увеличиваем CSG `{R4, R5}` до `{R4, R5, R6}`
11. Переходим к `R3` (множество исключенных `{R1, R2}`), но для него соседей нет, так как:

>     - Все гиперузлы с которыми есть соединение через простые гиперреба (`R1`, `R2`) лежат в множестве исключенных

Reviewer: ```suggestion
    - Все гиперузлы, с которыми есть соединение через простые гиперребра (`R1`, `R2`), лежат во множестве исключенных.
```

>     - Представитель левого вершины единственного сложного гиперребра (`min({R1, R2, R3}) = R1`) тоже лежит в множестве исключенных

Reviewer: Тут что-то не вяжется в "левого вершины", может быть "левой вершины"

12. Переходим к `R2`
13. Для него есть сосед `R3`, для которого создаем CMP
14. Увеличиваем сам CSG до `{R2, R3}`
15. Переходим к `R1`
16. Для него есть сосед `{R2}`, для которого и создаем CMP
17. Для самого `R2` есть сосед `R3` - увеличиваем CMP
18. Увеличиваем исходный подграф `{R1}` до `{R1, R2}` (сосед)
19. Для получившегося подграфа есть сосед `R3` (вывели из соседства с `R2`)
20. Увеличиваем подграф `{R1, R2}` до `{R1, R2, R3}` с помощью этого соседа

> 21. Используем единственное сложное гиперребро и получаем соседа для `{R1, R2, R3}` - это `{R4}`, представитель правой вершины `{R4, R5, R6}`. Для него создали CMP

Reviewer: ```suggestion
21. Используем единственное сложное гиперребро и получаем соседа для `{R1, R2, R3}` — это `{R4}`, представитель правой вершины `{R4, R5, R6}`. Создаем для него CMP.
```

22. Увеличиваем CMP до `{R4, R5}` с помощью простого соседа
23. Еще раз увеличиваем CMP, уже до `{R4, R5, R6}`

> 24. Возвращаемся к 20 и начинаем увеличивать подграф. Вначале добавляем `{R4}`

Reviewer: ```suggestion
24. Возвращаемся к шагу 20 и начинаем увеличивать подграф: добавляем `{R4}`.
```

> 25. Затем `{R5}`

Reviewer: В каждом пункте списка должно быть действие. А вначале и затем не нужны, так как это и так последовательность действий.
```suggestion
25. Добавляем `{R5}`.
```

> 26. В конце `{R6}`. Теперь CSG равен всему гиперграфу и мы можем закончить

Reviewer: ```suggestion
26. Наконец добавляем `{R6}`. Теперь CSG равен всему гиперграфу и мы можем закончить.
```

> Алгоритм хороший, так почему бы не добавить его в PostgreSQL? Можно? Можно и всегда было! Для это и существует `join_search_hook` - точка, позволяющая подменить алгоритм перебора JOIN'ов.

Reviewer: ```suggestion
Алгоритм хороший, так почему бы не добавить его в PostgreSQL? Можно? Да, и всегда было можно! Для это и существует `join_search_hook` — точка, позволяющая подменить алгоритм перебора JOIN'ов.
```

## Расширение pg_dphyp

> Идея для создания этого расширения пришла ко мне спонтанно. Я изучал разные алгоритмы JOIN'ов и наткнулся на него. Немного поискав в интернете я не нашел ничего для PostgreSQL и понял, что надо сделать самому. Кому уже сразу хочется посмотреть на то, что получилось [вот ссылка на репозиторий](https://github.com/ashenBlade/pg_dphyp).

Reviewer: ```suggestion
Идея для создания этого расширения пришла ко мне спонтанно. Я изучал разные алгоритмы JOIN'ов и наткнулся на него. Немного поискав в интернете, я не нашел ничего для PostgreSQL и понял, что надо сделать самому. Кому уже сразу хочется посмотреть на то, что получилось [вот ссылка на репозиторий](https://github.com/ashenBlade/pg_dphyp).
```

> В контексте ядра работы алгоритма я не принес ничего нового, даже наоборот - больше копипастил у других. Перед тем как приступить к реализации своего, я посмотрел реализации нескольких СУБД. В частности посмотрел на YDB, MySQL и DuckDB. Если кто хочет изучить DPhyp по коду, то рекомендую посмотреть на [код YDB](https://github.com/ydb-platform/ydb/blob/c23202bc294cf703741f1ea6ac30786578a58920/ydb/library/yql/dq/opt/dq_opt_dphyp_solver.h) - код чистый и понятный, очень легко читать. Но сам я начал не с YDB, а с MySQL, чего не рекомендую, так как их реализация значительно изменена и оптимизирована - сразу разобраться не получится (только по комментариям и с изначальным пониманием самого DPhyp)

Reviewer: ```suggestion
В контексте ядра работы алгоритма я не принес ничего нового, даже наоборот — больше копипастил у других. Перед тем как приступить к реализации своего, я посмотрел реализации нескольких СУБД. В частности посмотрел на YDB, MySQL и DuckDB. Если кто-то хочет изучить DPhyp по коду, то рекомендую посмотреть на [код YDB](https://github.com/ydb-platform/ydb/blob/c23202bc294cf703741f1ea6ac30786578a58920/ydb/library/yql/dq/opt/dq_opt_dphyp_solver.h) - код чистый и понятный, очень легко читать. Но сам я начал не с YDB, а с MySQL, чего не рекомендую, так как их реализация значительно изменена и оптимизирована — сразу разобраться не получится (только по комментариям и с изначальным пониманием самого DPhyp).
```

> В реализации я старался быть ближе к статье и делать минимальное количество изменений. Например, название функций и некоторых переменных тоже самое, что и в статье. Но хоть изменений в ядре алгоритма практически нет, они есть на уровне принятия операционных решений. Первое решение, которое надо принять - это представление множеств.

Reviewer: ```suggestion
В реализации я старался быть ближе к статье и делать минимальное количество изменений. Например, название функций и некоторых переменных такие же, что и в статье. Но хоть изменений в ядре алгоритма практически нет, они есть на уровне принятия операционных решений. Первое решение, которое надо принять — это представление множеств.
```

### Представление множеств

Множества являются лошадкой алгоритма, поэтому эффективность всей программы зависит от эффективности этого атома.

> Разные СУБД делают это по разному. Например, DuckDB [использует числа напрямую и хранит их в массиве](https://github.com/duckdb/duckdb/blob/73f85abbbdd38555ef7afa08090dfb4b10120df8/src/include/duckdb/optimizer/join_order/join_relation.hpp#L24), [YDB - `std::bitset<>`](https://github.com/ydb-platform/ydb/blob/c23202bc294cf703741f1ea6ac30786578a58920/ydb/library/yql/dq/opt/dq_opt_join_cost_based.cpp#L341), а [MySQL простое 8-байтное число](https://github.com/mysql/mysql-server/blob/ff05628a530696bc6851ba6540ac250c7a059aa7/sql/join_optimizer/node_map.h#L40).

Reviewer: ```suggestion
Разные СУБД делают это по разному. Например:
- DuckDB использует [числа напрямую и хранит их в массиве](https://github.com/duckdb/duckdb/blob/73f85abbbdd38555ef7afa08090dfb4b10120df8/src/include/duckdb/optimizer/join_order/join_relation.hpp#L24).
- YDB — [`std::bitset<>`](https://github.com/ydb-platform/ydb/blob/c23202bc294cf703741f1ea6ac30786578a58920/ydb/library/yql/dq/opt/dq_opt_join_cost_based.cpp#L341).
- MySQL — [простое 8-байтное число](https://github.com/mysql/mysql-server/blob/ff05628a530696bc6851ba6540ac250c7a059aa7/sql/join_optimizer/node_map.h#L40).
```

> Кто работает с PostgreSQL знает, что там имеется [своя реализация множества - `Bitmapset`](https://github.com/postgres/postgres/blob/62a47aea1d8d8ea36e63fe6dd3d9891452a3f968/src/include/nodes/bitmapset.h#L49). Она используется повсеместно и часто для хранения ID отношений. Вроде бы бери раз дают, но проблема в том, что операций над множествами огромное количество, а `Bitmapset` создает новую копию каждый раз когда меняется, то есть это лишние аллокации памяти. В PostgreSQL эта проблема часто не возникает, так как после создания `Bitmapset` он редко меняется, но в моем случае это критично.

Reviewer: ```suggestion
Кто работает с PostgreSQL знает, что там имеется своя реализация множества — [`Bitmapset`](https://github.com/postgres/postgres/blob/62a47aea1d8d8ea36e63fe6dd3d9891452a3f968/src/include/nodes/bitmapset.h#L49). Она используется повсеместно и часто для хранения ID отношений. Вроде бы бери раз дают, но проблема в том, что операций над множествами огромное количество, а `Bitmapset` создает новую копию каждый раз когда меняется, то есть это лишние аллокации памяти. В PostgreSQL эта проблема часто не возникает, так как после создания `Bitmapset` он редко меняется, но не в моем случае и для меня это критично.
```

> Проблему я решил реализовав сразу 2 подхода - создал 2 файла где в одном использовал `bitmapword` (8 байтное число, как в MySQL), а в другом `Bitmapset` (для сложных запросов). Но это произошло в самом начале разработки, когда я еще не особо понимал как работает алгоритм и его тонкости, поэтому через некоторое время на файл с `Bitmapset` забил (решил добавить изменения позже), а потом вообще удалил. В итоге, сейчас я использую представление множества с помощью числа.

Reviewer: ```suggestion
Проблему я решил реализовав сразу два подхода — создал два файла, где в одном использовал `bitmapword` (8 байтное число, как в MySQL), а в другом `Bitmapset` (для сложных запросов). Но это произошло в самом начале разработки, когда я еще не особо понимал, как работает алгоритм и его тонкости, поэтому через некоторое время на файл с `Bitmapset` забил (решил добавить изменения позже), а потом вообще удалил. В итоге, сейчас я использую представление множества с помощью числа.
```

> Проблем это не доставляет. Основные операции с множеством выполняются простыми битовыми `|`, `&` и `~`. Но есть и еще пара операций, которая важна для самого алгоритма. Например, итерирование по элементам множества (например, в процессе вычисления соседей). Таких операций много, поэтому я вынес их в отдельный [заголовочный файл](https://github.com/ashenBlade/pg_dphyp/blob/0cdc5b410d3bce41398a6646c576cca77994b6e3/simplebms.h).

Reviewer: ```suggestion
Проблем это не доставляет. Основные операции со множеством выполняются простыми битовыми `|`, `&` и `~`. Но есть и еще пара операций, которая важна для самого алгоритма. Например, итерирование по элементам множества (например, в процессе вычисления соседей). Таких операций много, поэтому я вынес их в отдельный [заголовочный файл](https://github.com/ashenBlade/pg_dphyp/blob/0cdc5b410d3bce41398a6646c576cca77994b6e3/simplebms.h).
```

Другая интересная операция - итерирование по всем подмножествам. Это нужно для расширения csg/cmp. Так как множество это число, то и операция проводится с числом. В MySQL это решили с помощью битового трюка `(init - state) & state`, который при постоянном применении ведет себя как инкремент, но при этом изменяются только биты множества. Эту реализацию я [взял себе](https://github.com/ashenBlade/pg_dphyp/blob/0cdc5b410d3bce41398a6646c576cca77994b6e3/pg_dphyp.c#L584).

Например, для множества `01010010` мы получим следующие подмножества:

```text
00000010    001
00010000    010
00010010    011
01000000    100
01000010    101
01010000    110
01010010    111
```

Слева само множество, а справа его битовая маска. Так как мы итерируемся инкрементом, то можно сказать, что эта битовая маска в числовом представлении также отображает и номер итерации. Дальше, это свойство будет использоваться часто.

### DP таблица

> DPhyp - это алгоритм динамического программирования и у него есть своя таблица для отслеживания состояния выполнения.

Reviewer: ```suggestion
DPhyp — это алгоритм динамического программирования, и у него есть своя таблица для отслеживания состояния выполнения.
```

> Если посмотреть на алгоритм из статьи, то видно, что эта таблица используется только для хранения готовых планов. В PostgreSQL для готовых планов (представление отношения) используется структура `RelOptInfo` и *она уже хранится в хэш-таблице*. Вроде вот и хорошо - не надо думать о создании своей, но нет.

Reviewer: ```suggestion
Если посмотреть на алгоритм из статьи, то видно, что эта таблица используется только для хранения готовых планов. В PostgreSQL для готовых планов (представлений отношений) используется структура `RelOptInfo` и *она уже хранится в хэш-таблице*. Вроде вот и хорошо — не надо думать о создании своей, но нет.
```

> Проблема заключается уже в самом PostgreSQL, а точнее его обработке `FULL JOIN`. Для этого типа соединений сейчас поддерживается только предикат равенства, а в коде, когда встречается такой предикат, все отношения в левой и правой части попадают в отдельные списки, которые планируются *независимо*. Это причина по которой, для внутренних таблиц используется своя система индексации (то есть индексы узлов DPhyp не обязательно соответствуют индексам отношений). Поэтому даже если я буду на время превращать `bitmapword` в `Bitmapset` (что потребует еще выделения памяти), то не смогу этого сделать, если индексы отношений больше 64 (максимального значения для 8-байтного числа).

Reviewer: ```suggestion
Проблема заключается уже в самом PostgreSQL, а точнее его обработке `FULL JOIN`. Для этого типа соединений сейчас поддерживается только предикат равенства, а в коде, когда встречается такой предикат, все отношения в левой и правой части попадают в отдельные списки, которые планируются *независимо*. Это причина по которой, для внутренних таблиц используется своя система индексации (то есть индексы узлов DPhyp не обязательно соответствуют индексам отношений). Поэтому даже если я захочу на время превращать `bitmapword` в `Bitmapset` (что потребует еще выделения памяти), то не смогу этого сделать, если индексы отношений больше 64 (максимального значения для 8-байтного числа).
```

Поэтому для алгоритма используется своя DP таблица. По факту, это хэш-таблица - `HTAB *`, часть PostgreSQL. Особенность этой таблицы в том, что в ней хранится только 1 структура, а ее ключ элемента должен хранится в начале этой структуры. Из-за этого была добавлена еще 1 структура `HyperNode`, которая представляет гиперузел. Но сейчас это не просто пара из множества узлов и отношения - в ней хранятся и другие данные.

### Построение графа

Другая не менее важная задача - это построение самого графа. Проблема в том, что в отличие от YDB или MySQL, балом я не правлю и должен подстраиваться под саму БД.

С одной стороны, проблемы вроде нет - я могу пройтись по всем предикатам, используемым в запросе и из них создать ребра. Собственно, сейчас это так и реализуется. Но дьявол кроется в деталях.

> Для начала - эту информацию мне приходится брать из 3 различных мест:

Reviewer: ```suggestion
Информацию для этой операции мне приходится брать из 3 различных мест:
```

1. `RelOptInfo->joinclauses` - список из предикатов, которые используют больше 1 отношения, то есть по факту это JOIN условие.

> 2. `PlannerInfo->join_info_list` - список из non-INNER JOIN условий

Reviewer: ```suggestion
1. `PlannerInfo->join_info_list` — список из non-INNER JOIN условий.
```

> 3. `PlannerInfo->eq_classes` - список классов эквивалентности (далее)

Reviewer: Зачем здесь "(далее)"? Может убрать?
```suggestion
1. `PlannerInfo->eq_classes` — список классов эквивалентности (далее).
```

Самое простое - это 2. Этот список содержит ограничения, которые накладывают разные не INNER (т.е. LEFT/RIGHT/FULL и т.д.) JOIN'ы. В нем есть 2 пары частей: синтаксические ограничения и минимальные (нужны для непосредственного вычисления). На всякий случай, я создаю гиперребро для обеих пар.

Про остальные тоже надо сказать. Первое имеет сложности с точки зрения *самого выражения* - оно может и не быть бинарным, а может быть, но по обеим сторонам используется одно и то же отношение. Для таких моментов я добавил свое понятие - cross join set (cjs). По факту, это просто множество отношений, которые должны соединиться между собой. Для каждого отношения в cjs я создаю простые ребра (каждый с каждым). Это решает проблему того, что какие-то предикаты могут отсутствовать. DPsize (который по умолчанию в PostgreSQL) решает это за счет того, что обрабатывает каждую возможную пару.

> Теперь 3 - классы эквивалентности. Класс эквивалентности - это механизм PostgreSQL, с помощью которого он определяет, что какие-то выражения равны друг другу. Это используется не только в выражениях. Например, выражения в `ORDER BY` или `GROUP BY` представляются именно классом эквивалентности (возможно, вырожденным, с единственным элементом). Но сейчас не об этом, а том как это стреляет. Такие классы эквивалентности появляются, когда в запросе имеются выражения равенства, даже одного достаточно для создания класса эквивалентности. Как можно догадаться, они также используются для JOIN'ов. Когда я встречаю класс эквивалентности с несколькими отношениями, то приходится создавать гиперребра для каждой пары. Поэтому подобные запросы с одними равенствам под капотом превращается в клику.

Reviewer: Что такое "клику"? Есть синоним попонятней?
```suggestion
Теперь 3 — классы эквивалентности. Класс эквивалентности — это механизм PostgreSQL, с помощью которого он определяет, что какие-то выражения равны друг другу. Это используется не только в выражениях. Например, выражения в `ORDER BY` или `GROUP BY` представляются именно классом эквивалентности (возможно вырожденным, с единственным элементом). Но сейчас не об этом, а том, как это стреляет. Такие классы эквивалентности появляются, когда в запросе имеются выражения равенства, даже одного достаточно для создания класса эквивалентности. Как можно догадаться, они также используются для JOIN'ов. Когда я встречаю класс эквивалентности с несколькими отношениями, то приходится создавать гиперребра для каждой пары. Поэтому подобные запросы с одними равенствам под капотом превращаются в клику.
```

Для примера, имеется такой запрос:

```sql
SELECT * FROM t1 
         JOIN t2 ON t1.x = t2.x AND t1.y > t2.y
         JOIN t3 ON t1.x + t2.x = t3.x
    LEFT JOIN t4 ON t4.x = t3.x;
```

Он имеет:

- Классы эквивалентности `{t1.x, t2.x}` и `{t1.x + t2.x, t3.x}`
- Предикат JOIN `t1.y > t2.y`
- Особый JOIN предикат `t4.x = t3.x`

Из них будут созданы следующие гиперребра:

- `{t1} - {t2}`
- `{t3} - {t1, t2}`
- `{t3} - {t4}`

### Несвязные подграфы

> Из предыдущего вытекает еще одна проблема. Что будет если мы получили несвязный граф (то есть лес)? В таком случае алгоритм ничего не сможет сделать. Точнее, он создаст план для каждого графа в лесу, но для *всего* запроса не сможет.

Reviewer: ```suggestion
Из выше сказанного вытекает еще одна проблема. Что будет, если мы получили несвязный граф (то есть лес)? В таком случае алгоритм ничего не сможет сделать. Точнее, он создаст план для каждого графа в лесу, но для *всего* запроса не сможет.
```

> Причем эта проблема может вылиться не только из-за `CROSS JOIN` или `,`, но и из-за внешних параметров подзапросов. Приведу пример:

Reviewer: ```suggestion
Причем эта проблема может появиться не только из-за `CROSS JOIN` или `,`, но и из-за внешних параметров подзапросов. Приведу пример:
```

```sql
SELECT * FROM t1
WHERE t1.x IN 
    (SELECT t2.x FROM t2, t3 WHERE t2.x = t1.x AND t3.x = t1.x);
```

В этом подзапросе `t1.x` является параметром, но на выходе в графе у нас лес, потому что для `t1` в подзапросе нет индекса. Можно представить и более сложные запросы, которые могут упасть таким образом. Вообще, это я обнаружил, когда запустил `\d` (чтобы отобразить все таблицы) в `psql` и оказалось, что там был примерный запрос.

> С одной стороны, на практике несвязные графы довольно редкое явление, чтобы тратить ресурсы для обнаружения этого, с другой, из-за подобного мы можем потратить лишнее время (просто удвоить время планирования без полезной нагрузки). Поэтому решение как поступать я отдал на откуп пользователям.

Reviewer: ```suggestion
С одной стороны на практике несвязные графы довольно редкое явление, чтобы тратить ресурсы для обнаружения этого, с другой из-за подобного мы можем потратить лишнее время (просто удвоить время планирования без полезной нагрузки). Поэтому решение, как поступать, я отдал на откуп пользователям.
```

> Имеется отдельный параметр `pg_dphyp.cj_strategy`, который принимает 3 значения:

Reviewer: ```suggestion
Для этого имеется отдельный параметр `pg_dphyp.cj_strategy`, который принимает 3 значения:
```

- `no` - если не смогли построить план, то вызвать DPsize/GEQO
- `pass` - если не смогли построить план, то собрать все готовые планы несвязных деревьев и отдать DPsize/GEQO на финальную доработку
- `detect` - создавать гиперребра для несвязных графов перед началом работы

Кажется, что первый вариант лишний, но на самом деле не особо, так как планы, созданные для этих подграфов могу быть не оптимальными из-за причины выше (между несвязными подграфами могут быть неявные связи, которые помогают создать оптимальный план), поэтому и финальный план тоже может быть не оптимальным.

> Остается вопрос - как находить несвязные графы. Ответ на поверхности - [Union-Set алгоритм](https://github.com/ashenBlade/pg_dphyp/blob/0cdc5b410d3bce41398a6646c576cca77994b6e3/pg_dphyp.c#L975). Для производительности  используется оптимизированная версия - с ранжированием и обновлением лидера у обеих частей.

Reviewer: ```suggestion
Остается вопрос — как находить несвязные графы? Ответ на поверхности — [алгоритм Union-Set](https://github.com/ashenBlade/pg_dphyp/blob/0cdc5b410d3bce41398a6646c576cca77994b6e3/pg_dphyp.c#L975). Для производительности используется оптимизированная версия — с ранжированием и обновлением лидера у обеих частей.
```

### Хранение гиперребер

Еще одна задача - хранение гиперграфа. Граф можно хранить в виде списка ребер, а можно таблицей смежности. Но для гиперграфа, остается только первый вариант, так как практически невозможно построить таблицу смежности для гиперребер (это не 1 элемент, а целое множество).

> Хорошо, определились, но можно ли применить какие-нибудь оптимизации? Да, можно. Почти у всех реализаций (например, [YDB](https://github.com/ydb-platform/ydb/blob/c23202bc294cf703741f1ea6ac30786578a58920/ydb/library/yql/dq/opt/dq_opt_join_hypergraph.h#L84) и [MySQL](https://github.com/mysql/mysql-server/blob/ff05628a530696bc6851ba6540ac250c7a059aa7/sql/join_optimizer/hypergraph.h#L69)) есть оптимизация для простых гиперребер. Напомню, что у простого гиперребра по обеим сторонам множества из единственного элемента. Эта оптимизация заключается в том, что мы храним все узлы, с которыми нас связывают простые гиперребра, в одном единственном множестве. Далее, в процессе работы нам нужно сделать единственную операцию для проверки сразу нескольких ребер. Такое множество часто называют `simple neighborhood`, почему, наверное, понятно.

Reviewer: ```suggestion
Хорошо, определились, но можно ли применить какие-нибудь оптимизации? Да, можно. Почти у всех реализаций (например, [YDB](https://github.com/ydb-platform/ydb/blob/c23202bc294cf703741f1ea6ac30786578a58920/ydb/library/yql/dq/opt/dq_opt_join_hypergraph.h#L84) и [MySQL](https://github.com/mysql/mysql-server/blob/ff05628a530696bc6851ba6540ac250c7a059aa7/sql/join_optimizer/hypergraph.h#L69)) есть оптимизация для простых гиперребер. Напомню, что у простого гиперребра по обеим сторонам находятся множества из единственного элемента. Эта оптимизация заключается в том, что мы храним все узлы, с которыми нас связывают простые гиперребра, в одном единственном множестве. Далее в процессе работы нам нужно сделать единственную операцию для проверки сразу нескольких ребер. Такое множество часто называют `simple neighborhood`, почему, наверное, понятно.
```

> Я тоже взял себе эту оптимизацию, но пошел чуть дальше и храню этот `simple neighborhood` для каждого гиперузла (в структуре `HyperNode`). Это немного упрощает работу, так как не нужно тратить время на дополнительное итерирование по узлам, а места занимает только 8 байт.

Reviewer: ```suggestion
Я тоже взял себе эту оптимизацию, но пошел чуть дальше и храню этот `simple neighborhood` для каждого гиперузла (в структуре `HyperNode`). Это немного упрощает работу, так как не нужно тратить время на дополнительное итерирование по узлам, и места он занимает всего 8 байт.
```

> А что по сложным гиперребрам? Тут тоже не без замечаний. Ранее я сказал, что для одного и того же выражения могут существовать несколько выражений, но с разным набором индексов необходимых отношений. Тогда же я и сказал, что не могу эти дополнительные индексы обработать. Поэтому в результате я могу получить множество дубликатов.

Reviewer: Можно перефразировать "**выражения** могут существовать несколько **выражений**", а то повторяется и непонятно?

С этим я решил бороться обычной [сортировкой](https://github.com/ashenBlade/pg_dphyp/blob/0cdc5b410d3bce41398a6646c576cca77994b6e3/pg_dphyp.c#L1096) - все ребра я храню в отсортированном виде и если при вставке нахожу равное ребро, то ничего не делаю (не вставляю).

### Создание плана

Сама статья не только про алгоритм на гиперграфах, но также и про эффективное создание плана. Это важный момент, так как некоторые JOIN операторы не коммутативны (например, `LEFT JOIN`), и если это не учитывать, возможно придется вычислять одну и ту же пару дважды.

Но это еще не все. Напомню, что DPsize плотно засел в кодовой базе, да на столько, что нельзя создать план для `i` таблиц, без нахождения оптимального плана для нижележащих.

Изначально я так и делал - после создания плана вызывал `set_cheapest` для нахождения лучшего плана и дальше мог спокойно вызывать `make_join_rel` для других. Но если посмотреть внутрь, то увидим, что она проходит по всем найденным путям для нахождения лучшего, то есть по мере работы программы время, затраченное на ее выполнение, будет только увеличиться.

> Из-за этого я перешел на DPsize-like подход. Для каждого гиперузла я веду список пар гиперузлов-кандидатов, которые могут его создать, а в конце рекурсивно вызываю создание плана. Этакий pull подход. Да, есть вероятность, что внутри окажется пара, которая не сможет создать план запроса и я потрачу время впустую, но накладные расходы на изначальный подход (с постоянным вызовом `set_cheapest`) все же больше, а, тем более, вероятность того, что какая-то пара гиперузлов не создаст полезного JOIN'а крайне мала из-за самой идеи DPhyp.

Reviewer: ```suggestion
Из-за этого я перешел на DPsize-like подход. Для каждого гиперузла я веду список пар гиперузлов-кандидатов, которые могут его создать, а в конце рекурсивно вызываю создание плана. Этакий pull подход. Да, есть вероятность, что внутри окажется пара, которая не сможет создать план запроса и я потрачу время впустую, но накладные расходы на изначальный подход (с постоянным вызовом `set_cheapest`) все же больше, а вероятность того, что какая-то пара гиперузлов не создаст полезного JOIN'а крайне мала из-за самой идеи DPhyp.
```

## Оптимальное итерирование по соседям

> Наконец-то мы пришли к самой интересной части - как мы будем обходить соседей. В алгоритме можно выделить 3 функции, которые заняты перебором пар csg/cmp и в них нам нужно вычислять соседей целых 4 раза. Все становится страшнее, когда осознаешь, что алгоритм предполагает итерирование по всем возможным подмножествам соседей ([power set](https://en.wikipedia.org/wiki/Power_set) без пустого множества) - это $2^i$ комбинаций.

Reviewer: Ты случайно вместо `$2^i$` не хотел написать ``` `2^i` ```?
В следующей строке и ниже аналогично.

Имеет ли смысл оптимизировать это место? В коде MySQL довольно много комментариев, которыми они описывают те или иные принятые решения (чаще всего на основании микробенчмарков). Один из [таких комментариев](https://github.com/mysql/mysql-server/blob/ff05628a530696bc6851ba6540ac250c7a059aa7/sql/join_optimizer/subgraph_enumeration.h#L280) написан к функции вычисления соседей `FindNeighborhood`:

> ...
> This function accounts for roughly 20–70% of the total DPhyp running time, depending on the shape of the graph (~40% average across the microbenchmarks)
> ...

То есть практически все время работы алгоритма занимает логика вычисления соседей, поэтому - да, смысл в ее оптимизации есть.

Для того, чтобы ускорить вычисление соседей в MySQL используют кэширование по подмножествам.

### Подход в MySQL

Их идея построена на свойстве подмножеств, которые мы получаем при итерировании инкрементом.

> Если посмотреть на этот способ повнимательнее, то можно заметить, что при таком подходе при следующем шаге мы часть получаем подмножество предыдущего шага. И при этом очень часто (каждый следующий шаг) первый бит постоянно переключается то в `0`, то в `1`. А если его переключить из `0` в `1`, то мы получим подмножество! Например, `0011100` -> `0011101`.

Reviewer: Тут что-то не вяжется в "мы часть получаем", надо поправить.

Сразу возникает желание просто взять предыдущих соседей и добавить этим первым элементом, но возникает вопрос - а корректно ли это? Да, корректно. Если еще раз взглянем на определение соседей из статьи, то не увидим там ограничение на порядок обхода узлов, а это значит, что мы можем взять одно множество и дополнить еще одним.

> Чтобы стала понятна идея рассмотрим на примере итерации по подмножеству из 4 узлов. Привожу пример из комментария:

Reviewer: ```suggestion
Чтобы стала понятна идея, рассмотрим на примере итерации по подмножеству из 4 узлов. Привожу пример из комментария:
```

```text
00000001   001
00001000   010
00001001   011
00100000   100
00100001   101
00101000   110
00101001   111
```

<spoiler title="Ограничене схемы кеширования">

Схема кэширования соседей работает только в случае, когда множество исключенных узлов зафиксировано. Под это условие попадают последние циклы в `Enumerate*` функциях, когда множество исключенных добавляется соседями csg/cmp (в определении из статьи эти функции рекурсивно вызывают себя, но с множеством исключенных узлов $X \cup \mathcal{N}(S, X)$, которое не зависит от цикла).

Но даже этого достаточно для прироста производительности, так как остальные 2 места перебора подмножеств:

- `EnumerateCmpRecursive`, но там вызывается `EmitCsgCmp` для которого находить соседей не нужно
- `EnumerateCsgRecursive`, но в нем вызывается `EmitCsg`, который сбрасывает множество исключенных узлов.

> Чтобы понять, почему нельзя кешировать соседей, если множество исключенных узлов меняется рассмотрим конкретный пример. Вызван `EnumerateCsgRecursive` с множеством $S = 000110$ и $X = 000111$. Для него есть множество соседей, по подмножествам которых будем итерироваться, $\mathcal{N} = 1110000$ (то есть получается, что $0001000$ пока свободен и никому не принадлежит).

Reviewer: ```suggestion
Чтобы понять, почему нельзя кешировать соседей, если множество исключенных узлов меняется, рассмотрим конкретный пример. Вызван `EnumerateCsgRecursive` со множеством $S = 000110$ и $X = 000111$. Для него есть множество соседей, по подмножествам которых будем итерироваться, $\mathcal{N} = 1110000$ (то есть получается, что $0001000$ пока свободен и никому не принадлежит).
```

> Мы хотим эффективно вычислять соседей и для этого кешируем их. Возникает вопрос - что кешировать? Неизменяемая часть - это $S$, к которой добавляется подмножество соседей. Решаем закешировать его, но тут возникает такая ситуация: для какого-то узла есть гиперребро правая часть которого $1001000$. Как это ребро корректно обработать? Имеем 2 возможных исхода:

Reviewer: ```suggestion
Мы хотим эффективно вычислять соседей и для этого кешируем их. Возникает вопрос — что кешировать? Неизменяемая часть — это $S$, к которой добавляется подмножество соседей. Решаем закешировать его, но тут возникает такая ситуация: для какого-то узла есть гиперребро, правая часть которого $1001000$. Как это ребро корректно обработать? Имеем два возможных исхода:
```

> 1. В множество исключенных узлов *не были* добавлены все соседи. Тогда в соседи для `EmitCsg` добавляется узел $0001000$, который будет прокидываться везде. Но это значит, что для итераций, в которых *участвует* $1000000$ будет нарушена логика вычисления соседей в `EmitCsg`, так как этот элемент будет в исключенном множестве, а соответственно *ребро не должно учитываться*

Reviewer: Эти пункты должны быть ненумерованного списка, здесь же два равнозначных варианта, а не последовательность действий.

2. В множество исключенных узлов *были* добавлены все соседи. Тогда мы получаем обратную сторону медали - в итерациях, где элемент $1000000$ *не участвует*, *будет отсутствовать какой-то узел*

Но если посмотреть в код MySQL, то можно увидеть, что они все же используют кеширование. Но почему? Если говорить просто, то они вычисляют соседей для подмножества, а затем к получившимся соседям добавляют исключенные узлы и исходных соседей.

> Исключенные они добавляют по логике `Enumerate*` функций, так как к ним на вход передаются исключенные множества, которые содержат все ранее найденные множества, хотя рекурсивным вызовом могли передать только небольшое подмножество. Например, в процессе работы `EnumerateCsgRecursive` рекурсивно вызывал себя несколько раз и каждый раз множество соседей составляло 2 элемента, но рекурсивные вызовы передавали только 1 элемент, хотя аккумулированные исключенные узлы это объединение всех найденных соседей. То есть, множество исключенных узлов, которые нам передали может содержать все узлы, которые будут валидными соседями в `EmitCsg` (так как он сбрасывает текущие исключенные и начинает использовать свои).

Reviewer: ```suggestion
Исключенные они добавляют по логике `Enumerate*` функций, так как к ним на вход передаются исключенные множества, которые содержат все ранее найденные множества, хотя рекурсивным вызовом могли передать только небольшое подмножество. Например, в процессе работы `EnumerateCsgRecursive` рекурсивно вызывал себя несколько раз и каждый раз множество соседей составляло два элемента, но рекурсивные вызовы передавали только один элемент, хотя аккумулированные исключенные узлы это объединение всех найденных соседей. То есть, множество исключенных узлов, которые нам передали, может содержать все узлы, которые будут валидными соседями в `EmitCsg` (так как он сбрасывает текущие исключенные и начинает использовать свои).
```

Пример: после нескольких рекурсивных вызовов `EnumerateCsgRecursive` имеет $S = 1010001011$, $X = 1111111111$, но при вызове `EmitCsg` множество исключенных будет сброшено до $X = 0000000011$ (например, потому что наименьший узел это 2). Тогда добавление $X \\ S = 0101110100$ будет просто означать, что мы на всякий случай хотим обработать те узлы, которые были соседями предыдущих вызовов, но в текущее множество CSG не попали.

> Исходных соседей добавляют по той же самой логике. Но чтобы не нарушать корректность алгоритма, из полученного множества удаляются все узлы, которые находятся в полученном подграфе. Вкратце: авторы изначально знают, что соседи полученные таким способом могут не содержать всех соседей, поэтому добавляют вообще все, что может быть правдой, даже если это приведет к избыточным вычислениям.

Reviewer: ```suggestion
Исходных соседей добавляют по той же самой логике. Но чтобы не нарушать корректность алгоритма, из полученного множества удаляются все узлы, которые находятся в полученном подграфе. Если коротко: авторы изначально знают, что соседи полученные таким способом могут не содержать всех соседей, поэтому добавляют вообще все, что может быть правдой, даже если это приведет к избыточным вычислениям.
```

Как только я понял идею, то переписал этот код себе. Буквально. Переписал почти слово в слово, только ~~Васю на Петю~~ `taboo` на `forbidden` поменял (чтобы хоть что-то свое было). Код жил так довольно долго, несколько недель. А потом я спохватился - GPLv2!

Код MySQL распространяется под лицензией GPLv2, а учитывая, что я переписал почти все слово в слово (на тот момент, наверное, еще даже не до конца понимая саму идею), то нарушил эту лицензию - мой то проект под MIT, они не совместимы. Тогда я встал перед вопросом - выкинуть эту хорошую оптимизацию и сделать код медленным или оставить, но поменять лицензию на GPLv2. По итогу выбрал первое и это стало началом увлекательного думания на несколько недель - как мне оптимизировать эту комбинаторику.

### Суффиксный кэш

> Передо мной стоит задача - оптимизировать алгоритм, но так, чтобы это не было копиркой MySQL (в моем случае, это важный аспект, потому что после нахождения первого решения часто использую только его). Сама идея ясна - зачем вычислять все множество соседей, если можно просто его дополнять этой дельтой. Нам нужно как-то найти то-ли шаблон, то-ли еще что, что поможет нам обнаружить подобное подмножество.

Reviewer: ```suggestion
Передо мной стоит задача — оптимизировать алгоритм, но так, чтобы это не было копиркой MySQL (в моем случае это важный аспект, потому что после нахождения первого решения часто использую только его). Сама идея ясна — зачем вычислять все множество соседей, если можно просто его дополнять этой дельтой. Нам нужно как-то найти то ли шаблон, то ли еще что-нибудь, что поможет нам обнаружить подобное подмножество.
```

> А что если посмотреть с другой стороны? Действительно, у нашего способа итерирования по подмножествам есть хорошее свойство - верхняя часть (MSB) меняется гораздо реже чем нижняя (LSB). А почему-бы нам не использовать это свойство - просто закэшируем то, что редко меняется!

Reviewer: ```suggestion
А что если посмотреть с другой стороны? Действительно, у нашего способа итерирования по подмножествам есть хорошее свойство — верхняя часть (MSB) меняется гораздо реже, чем нижняя (LSB). А почему бы нам не использовать это свойство — просто закешируем то, что редко меняется!
```

Но что такое "редко меняется"? В первой идее за эту константу я взял замыкающие единицы - последовательность последних единиц (при фиксированном разряде) в числе может только увеличиваться, то есть достаточно только подсчитать соседей для этой последовательности, а затем добавлять меняющуюся. Отделить базовую (не меняющуюся) часть мы можем простой битовой маской, а ее размер мы знаем (подсчитали).

> > Если вспомнить оптимальную последовательность итерирования, предложенный MySQL, то можно заметить, что эта стратегия идеально подходит под такую последовательность.

Reviewer: ```suggestion
> Если вспомнить оптимальную последовательность итерирования, предложенную MySQL, то можно заметить, что эта стратегия идеально подходит под такую последовательность.
```

Как нам отследить изменение этой части? Не проблема - это ведь двоичная арифметика, просто отслеживаем сколько итераций до следующей новой `1`, а затем делим на 2 и ожидаем полученное количество итераций. Когда начинается новый разряд, то просто сбрасываем счетчик и повторяем.

Алгоритм следующий:

1. Инициализация:

>    1. `nb_cache = 0` Закешированные соседи

Reviewer: ```suggestion
   1. `nb_cache = 0` закешированные соседи.
```

>    2. `nb_cache_subset = 0` Маска множества, для которого хранится `nb_cache`

Reviewer: ```suggestion
   2. `nb_cache_subset = 0` маска множества, для которого хранится `nb_cache`.
```

>    3. `prev_update = 0` Сохраненное значение `next_update`

Reviewer: ```suggestion
   3. `prev_update = 0` сохраненное значение `next_update`.
```

>    4. `next_update = 0` Итераций до следующего обновления кэша

Reviewer: ```suggestion
   4. `next_update = 0` итераций до следующего обновления кеша.
```

2. Для очередной итерации (начиная с 1):
   1. Если `next_update == 0`:

>       1. Довычисляю соседей последним элементом

Reviewer: ```suggestion
      1. Довычисляю соседей последним элементом.
```

>       2. `next_update = prev_update / 2` - следующая единица в последовательности появится ровно через половину числа предыдущих операций

Reviewer: ```suggestion
      2. `next_update = prev_update / 2` — следующая единица в последовательности появится ровно через половину числа предыдущих операций.
```
Далее аналогично.

      3. `prev_update = next_update` - это нужно для вычисления следующих соседей
      4. `nb_cache = *вычисленные сейчас соседи*`
      5. `nb_cache_subset = *текущее подмножество*`
   2. Если в множестве только 1 элемент:
      1. Вычисляю соседей для этого единственного элемента
      2. `next_update = 2^*номер разряда - 2*` - столько осталось до добавления еще одной единицы в множество
      3. `prev_update = next_update`
      4. `nb_cache = *вычисленные сейчас соседи*`
      5. `nb_cache_subset = *текущее подмножество*`
   3. В противном случае:
      1. Из подмножества удаляем закешированные элементы (храним в `nb_cache_subset`)
      2. Вычисляем соседей с `nb_cache` в качестве базы

> ### 1-слойный кэш

Reviewer: Может лучше
```suggestion
### Однослойный кеш
```

> Хорошая ли это идея? Как подспорье для дальнейших рассуждений да, но на практике эта схема кэширования даст прирост только в конце, когда у нас почти все заполнено единицами (так как способа итерирования по оптимальным подмножествам не нашел, значит итерирование инкрементом).

Reviewer: ```suggestion
Хорошая ли это идея? Как подспорье для дальнейших рассуждений — да, но на практике эта схема кеширования даст прирост только в конце, когда у нас почти все заполнено единицами (так как способа итерирования по оптимальным подмножествам не нашел, значит итерирование инкрементом).
```

Вернемся в самое начало. Мы обсуждали то, что кешировать нужно редко меняющиеся части. За эту часть мы взяли последовательность замыкающих единиц. Она действительно редко меняется, но заметим, что эта часть вариативная, то есть надо отслеживать ее размер. Теперь сделаем эту часть фиксированной, то есть кэшировать будем определенный суффикс множества.

Вопрос - что даст больший прирост: кэширование замыкающих единиц или MSB? Давайте посчитаем. Для простоты сделаем это наглядно, на том же множестве из 4 элементов.

> Для схемы кэширования лидирующих `1` мы имеем следующую схему расчета (2 столбец количество обработанных узлов, а звездочкой пометил места обновления кэша):

Reviewer: ```suggestion
Для схемы кеширования лидирующих `1` мы имеем следующую схему расчета (второй столбец показывает количество обработанных узлов, а звездочкой отмечены места обновления кеша):
```

```text
0001  1
0010  1  *
0011  2
0100  1  *
0101  2
0110  2  *
0111  1
1000  1  *
1001  1
1010  1
1011  2
1100  2  *
1101  1
1110  3  *
1111  1 
```

> Всего надо обработать 22 элемента. На 2 больше чем схема с табу-битом.

Reviewer: ```suggestion
Всего надо обработать 22 элемента. На два больше, чем схема с табу-битом.
```

Со схемой кэширования MSB надо вначале ответить на вопрос - а что такое MSB, каков его размер? Возьмем мало, будем очень много вычислять, возьмем много будем часто вычислять этот MSB. В принципе, это все можно также вычислить, так как тут та же самая двоичная математика. Но для наглядности, возьмем MSB равный 2:

```text
0001  1
0010  1
0011  2
0100  1  *
0101  1
0110  1
0111  2
1000  1  *
1001  1
1010  1
1011  2
1100  2  *
1101  1
1110  1
1111  2
```

> Тут уже 20 элементов - столько же сколько, чем замыкающие единицы, но меньше чем замыкающие единицы.

Reviewer: Тут что-то несвязное, надо поправить.

> > Здесь каждый раз при обновлении кэша я честно полностью рассчитывал соседей. Но вы могли заметить, что при расчете закешированной части для соседей мы могли использовать ту же самую оптимизацию с последним битом - добавить к закешированным соседям этот последний элемент. Такая оптимизация позволит достичь всего-лишь 16 итераций, что намного эффективнее.

Reviewer: ```suggestion
> Здесь каждый раз при обновлении кеша я честно полностью рассчитывал соседей. Но вы могли заметить, что при расчете закешированной части для соседей мы могли использовать ту же самую оптимизацию с последним битом — добавить к закешированным соседям этот последний элемент. Такая оптимизация позволит достичь всего лишь 16 итераций, что намного эффективнее.
```

> Тогда я упустил этот момент и не стал рассматривать, но это было к лучшему и скоро узнаете почему.

Схема с кэшированием фиксированной части MSB показала себя лучше, да и к тому же у нее есть возможность настраивания (параметр - размер кешированной части). По итогу, выбираем схему кэширования с 1 элементом - MSB. Далее, эту неизменяемую часть, которую используем за стартовое значение для вычисления соседей, буду называть base (базой).

Алгоритм работы с ней довольно прост: через каждые $2^{len(64 - MSB)}$ элементов сохраняем в кэш эту базовую часть, а затем просто используем как стартовую точку при расчете соседей.

> ### 2-слойный кэш

Reviewer: Может лучше
```suggestion
### Двухслойный кэш
```

> Погодите, мы ведь работаем с множествами, а любое множество можно создать из предыдущего добавлением одного элемента! Например, множество `011010` можно построить из любого `001010`, `010010` или `011000`. Но это также значит, что можно создать текущее множество простым добавлением самого крайнего с начала элемента. Тот же изначальный пример с табу битом - это частный случай, когда мы добавляли первый элемент.

Reviewer: ```suggestion
Погодите, мы ведь работаем со множествами, а любое множество можно создать из предыдущего добавлением одного элемента! Например, множество `011010` можно построить из любого `001010`, `010010` или `011000`. Но это также значит, что можно создать текущее множество простым добавлением самого крайнего с начала элемента. Тот же изначальный пример с табу битом — это частный случай, когда мы добавляли первый элемент.
```

Давайте построим таблицу из 4 элементов и посмотрим, как мы можем создать текущее множество из предыдущего:

```text
0000 <+  <+  <+
   ^  |   |   |
0001  |   |   |
      |   |   |
0010 -+   |   |
   ^      |   |
0011      |   |
          |   |
0100 <+  -+   |
   ^  |       |
0101  |       |
      |       |
0110 -+       |
   ^          |
0111          |
              |
1000 <+  <+  -+
   ^  |   |
1001  |   |
      |   |
1010 -+   |
   ^      |
1011      |
          |
1100 <+  -+
   ^  |
1101  |
      |
1110 -+
   ^
1111
```

> Вы видите этот паттерн? Я да - мы берем элементы в ячейке под индексом меньше нашей на какую-то степень 2 и используем как базу для создания текущего множества добавлением текущего крайнего элемента. Что это за степень 2? Можно увидеть из этой же схемы - это количество текущих лидирующих нулей! То есть множество соседей, которое может использоваться для создания текущего располагается в $2^{zeros}$ шагах назад, где $zeros$ - это количество текущих лидирующих нулей номера итерации.

Reviewer: ```suggestion
Вы видите этот паттерн? Я да — мы берем элементы в ячейке под индексом меньше нашей на какую-то степень 2 и используем как базу для создания текущего множества добавлением текущего крайнего элемента. Что это за степень 2? Можно увидеть из этой же схемы - это количество текущих лидирующих нулей! То есть множество соседей, которое может использоваться для создания текущего, располагается в $2^{zeros}$ шагах назад, где $zeros$ — это количество текущих лидирующих нулей номера итерации.
```

| Размер таблицы | Занимаемая память | Всего элементов | Итераций | Сохранено |
| -------------- | ----------------- | --------------- | -------- | --------- |
| 2              | 32   б            | 4               | 3        | 1         |
| 3              | 64   б            | 12              | 7        | 5         |
| 4              | 128  б            | 32              | 15       | 17        |
| 5              | 256  б            | 80              | 31       | 49        |
| 6              | 512  б            | 192             | 63       | 129       |
| 7              | 1    Кб           | 448             | 127      | 321       |
| 8              | 2    Кб           | 1024            | 255      | 769       |
| 9              | 4    Кб           | 2304            | 511      | 1793      |
| 10             | 8    Кб           | 5120            | 1023     | 4097      |
| 11             | 16   Кб           | 11264           | 2047     | 9217      |
| 12             | 32   Кб           | 24576           | 4095     | 20481     |

Легенда:

> - Размер таблицы - размер множества, которое мы кэшируем

Reviewer: ```suggestion
- Размер таблицы — размер множества, которое мы кэшируем.
```

> - Занимаемая память - сколько занимает таблица для текущей реализации ($2^{table\ size + 3}$)

Reviewer: ```suggestion
- Занимаемая память — размер памяти, которое занимает таблица для текущей реализации ($2^{table\ size + 3}$).
```

> - Всего элементов - сумма элементов из всех подмножеств ($\sum_{k = 1}^{table\ size}k C_{table\ size}^{k}$)

Reviewer: ```suggestion
- Всего элементов — сумма элементов из всех подмножеств ($\sum_{k = 1}^{table\ size}k C_{table\ size}^{k}$).
```

> - Итераций - сколько должны обработать элементов в оптимизированной реализации, тоже самое что и количество итераций, но первую (пустую) не считаем ($2^{table\ size} - 1$)

Reviewer: ```suggestion
- Итераций — количество элементов , которые должны обработать в оптимизированной реализации, тоже самое что и количество итераций, но первую (пустую) не считаем ($2^{table\ size} - 1$).
```

> - Итераций сохранено - разница между "Всего элементов" и "Итераций"

Reviewer: ```suggestion
- Итераций сохранено — разница между "Всего элементов" и "Итераций".
```

> Какой размер таблицы использовать можно вычислять динамически или просто выбрать оптимальное значение под свою нагрузку эвристически.

Reviewer: ```suggestion
Какой размер таблицы использовать, можно вычислять динамически или просто выбрать оптимальное значение под свою нагрузку эвристически.
```

\sum_{k}^{max - tbl - 1}C_{max - tbl - 1}^{k}(2^{tbl} - 1 + k - 2^{tbl + 1} + 1 - k) + \sum_{k}^{max - tbl - 1}C_{max - tbl - 1}^{k - 1}(2^{tbl} - 1 + k) + 2^{tbl} - 1 + max - tbl = \\
-2^{tbl}\sum_{k}^{max - tbl - 1}C_{max - tbl - 1}^{k} + \sum_{k}^{max - tbl - 1}C_{max - tbl - 1}^{k - 1}(2 ^{tbl} - 1 + k) + 2^{tbl} - 1 + max - tbl = \\
2^{tbl}(\sum_{k}^{max - tbl - 1}C_{max - tbl - 1}^{k - 1} - \sum_{k}^{max - tbl - 1}C_{max - tbl - 1}^{k}) + \sum_{k}^{max - tbl - 1}C_{max - tbl - 1}^{k - 1}(2^{tbl} - 1 + k) + 2^{tbl} + max - tbl
$$

Если раскроем суммы под скобками в первом слагаемом, то получим сумму такого вида: $C^0 - C^1 + C^1 - C^2 + C^2 - ...$. Легко увидеть, что слагаемые взаимно уничтожаются и на выходе мы получаем $C^0 - C^{max - tbl - 1} = 0$ - первое слагаемое равно нулю. Значит итоговое выражение:

$$
\sum_{k}^{max - tbl - 1}C_{max - tbl - 1}^{k - 1}(2^{tbl} - 1 + k) + 2^{tbl} - 1 + max - tbl = \\
\sum_{k}^{max - tbl - 1}C_{max - tbl - 1}^{k - 1}(2^{tbl} - 1 + k) = \\
\sum_{k}^{max - tbl}C_{max - tbl - 1}^{k}(2^{tbl} + k)
$$

> Учитывая $max > tbl$ (так как формула предполагает, что мы увеличиваем $tbl$ на 1, то есть это предыдущее значение, а размер таблицы нет смысла увеличивать больше размера множество) видим, что это выражение неотрицательно, так как каждый член этой суммы неотрицателен.

Reviewer: ```suggestion
Учитывая $max > tbl$ (так как формула предполагает, что мы увеличиваем $tbl$ на 1, то есть это предыдущее значение, а размер таблицы нет смысла увеличивать больше размера множества) видим, что это выражение неотрицательно, так как каждый член этой суммы неотрицателен.
```

</spoiler>

> ### 3-слойный кэш

Reviewer: Может лучше
```suggestion
### Трехслойный кеш
```

С base частью изменений почти нет. Во-первых, как можно понять из предыдущих шагов, при начале новой base части нам нужно сбросить состояние table части (вычисление таблицы начинается заново), а также и hot части (quad leader выставить в текущее вычисленное значение соседей). Но и это еще не все - почему бы нам не использовать тот же трюк с нечетными итерациями? И верно же - если первый бит текущего номера итерации base части равен 1, то мы можем просто довычислить соседей, так же как делали раньше.

Последний вопрос - как это все комбинировать. Здесь надо использовать префикс. Легко увидеть, что:

- каждые $2^{table size}$ итераций обновлять base часть (что в терминах двоичных чисел означает, что первые $i$ битов 0)
- каждые $2^4 = 16$ итераций делать новые записи в таблицу
- каждые $8$ итераций обновлять quad leader в hot части

> Алгоритм значительно усложнился, но чего не сделаешь ради оптимизаций. Зато, теперь, для того же размера кэшируемого множества требуется не $2^{i}$ байт, а всего лишь $2^{i - 4}$. Например, вместо 8Кб теперь требуется только 0.5Кб, но при этом кэшируется все 10 элементов.

Reviewer: ```suggestion
Алгоритм значительно усложнился, но чего не сделаешь ради оптимизаций. Зато теперь для того же размера кешируемого множества требуется не $2^{i}$ байт, а всего лишь $2^{i - 4}$. Например, вместо 8Кб теперь требуется только 0.5Кб, но при этом кешируется все 10 элементов.
```

> Когда эта идея устаканилась в голове я уже принялся во всю писать код. И вдруг меня осенило.

Reviewer: ```suggestion
Когда эта идея устаканилась в голове, и я уже принялся во всю писать код, вдруг меня осенило.
```

Но как правильно ее использовать? Ответ тоже прост. Вспомним то, как мы вычисляем соседей - мы просто берем родительских соседей и дорасчитываем первым элементом. А вот и 2 наблюдение: родительское множество - это наше, но без крайнего бита, например, для `01110` родителем будет `01100`. А в какой ячейке хранятся соседи для `01100`? Правильно под индексом 2, т.к. у него 2 нуля в начале. Крайний случай когда мы переходим в новый старший разряд - тогда после удаления бита мы получим пустое множество, но для пустого множества мы возвращаем пустое множество, так как у него не может быть соседей.

А теперь, собственно, и сам алгоритм:

1. Создаем DP таблицу, (динамический) массив.

> 2. Начинаем очередную итерацию

Reviewer: И в остальных точки.
```suggestion
2. Начинаем очередную итерацию.
```

3. Из номера итерации убираем первый выставленный бит и считаем количество нулей
4. Из DP таблицы берем элемент под этим индексом
5. Рассчитываем соседей, беря за основу закешированных, а за обновление - первый элемент текущего подмножества
6. Сохраняем в таблицу полученное значение

Корректно ли это? Да. Проведем доказательство от обратного. Базой будет то, что соседи пустого множества - это пустое множество (либо другое, но определенное значение). Само доказательство исходит из свойства алгоритма перебора подмножеств - инкремента.

Доказательство:

1. Нам дана текущая итерация `iter`.

> 2. Мы убираем последний бит/элемент и получаем число `iter_parent`

Reviewer: И в остальных точки.
```suggestion
2. Мы убираем последний бит/элемент и получаем число `iter_parent`.
```

Утверждается, что `nb_parent` - это соседи множества `iter_parent`. Представим, что это не так, то есть в той ячейке хранятся соседи `iter_parent_2`, который не равен `iter_parent`. Тогда это число может быть либо больше, либо меньше его, но:

- `iter_parent_2` точно не может быть больше `iter_parent`, так как это означает, что он был бы больше и изначального `iter` (так как различия в MSB части), но мы итерируемся строго возрастающим способом, а значит этого числа мы еще не встречали и его значение в таблице быть не может
- числом меньше тоже быть не может, потому что это означает, что `iter_parent` мы пропустили: у обоих чисел одинаковое количество лидирующих нулей (по нашему предположению), а так как `iter_parent` идет *после* `iter_parent_2` (так как больше в числовом представлении), то значение под соответствующим индексом мы должны были перезаписать (перезаписать соседей `iter_parent_2`), но не сделали это, то есть нарушили алгоритм.

Единственный оставшийся вариант - под этим индексом хранятся соседи множества `iter_parent`. Учитывая, что для пустого множества соседи определены, мы также можем утверждать, что и мусора мы получить не можем:

> - если текущее множество состоит из 1 элемента, то значит мы перешли в новый старший разряд, в котором раньше не были раньше. Убрав этот элемент мы получим пустое множество, но для него соседи определены и для этого нового, ранее не виденного индекса соседи сохранятся и далее значение будет определено.

Reviewer: ```suggestion
- Если текущее множество состоит из 1 элемента, то значит мы перешли в новый старший разряд, в котором раньше не были. Убрав этот элемент мы получим пустое множество, но для него соседи определены и для этого нового, ранее не виденного индекса, соседи сохранятся и далее значение будет определено.
```

> - в противном случае, количество лидирующих нулей будет *не больше* номера старшего разряда (иначе мы начали новый разряд и это случай 1), и соседей для них мы уже должны были записать в таблицу, то есть оно определено.

Reviewer: ```suggestion
- В противном случае количество лидирующих нулей будет *не больше* номера старшего разряда (иначе мы начали новый разряд и это случай 1), и соседей для них мы уже должны были записать в таблицу, то есть оно определено.
```

    zero_count = bmw_rightmost_one_pos(iter_state->iteration);
    iter_state->cached_neighborhood[zero_count] = neighbors;

    return neighbors;
}
```

> Схему также можно чуть-чуть оптимизировать не сохраняя нечетные множества, но это уже область микро-оптимизаций.

А нельзя ли еще добавить оптимизаций? Да, можно. Пример этого уже есть в коде, который только что показал - индексирование.

<spoiler title="Не совсем идеальный кеш">

Да, эта стратегия кеширования хороша, но не идеальна. Вспомните начало, когда описывался подход кеширования MySQL - они использовали его даже в первом цикле `EnumerateCsgRec`. В нем множество исключенных узлов постоянно меняется и из-за этого в чистом виде кешировать нельзя. Они обошли это эвристикой - кэшируют исключенные узлы, а потом добавляют все ранее запрещенные, но которые *могли бы быть* соседями.

Пример, когда это *не* так. Есть 3 вызова `EnumerateCsgRec` (`S` - подграф, `X` - множество исключенных узлов, `N` - соседи, по которым сейчас происходит итерация, т.е. вычислены `N(S, X)`):

```text
1. S = 00000001, X = 00000001, N = 00001110
2. S = 00001001, X = 00001111, N = 01110000
3. S = 01001001, X = 01111111, N = 10000000
```

> Первые 2 итерации дали соседей из 3 рядом стоящих элементов и сейчас мы находимся в 3 вызове: подграф = `1001001`, исключенные = `1111111`, соседи = `10000000`. В первом (и единственном цикле) мы обнаружили, что у `11001001` есть план, поэтому нужно найти для этого множества соседей. Следуя логике MySQL, для него соседи - `10110110`, так как нужно добавить всех предыдущих соседей.

Reviewer: ```suggestion
Первые 2 итерации дали соседей из 3 рядом стоящих элементов, и сейчас мы находимся в 3 вызове: подграф = `1001001`, исключенные = `1111111`, соседи = `10000000`. В первом (и единственном цикле) мы обнаружили, что у `11001001` есть план, поэтому нужно найти для этого множества соседей. Следуя логике MySQL, для него соседи - `10110110`, так как нужно добавить всех предыдущих соседей.
```

Где здесь может возникнуть проблема? Посмотрите на вызов 2: откуда взялись соседи `1110000`? Мы получили их от `0000100`, но он *был не в нашем **итоговом** подграфе*, а значит, что одно из гиперребер могло содержать в себе `01000000`, который должен быть исключен (содержится в итоговом подграфе).

> Как не трудно догадаться, после этого количество узлов (а значит и возможных пар csg/cmp), которые следует рассмотреть, увеличится, а значит и количество "мусорных" решений, которые либо не дадут полезного ответа, либо будут рассмотрены несколько раз, тоже увеличиться. Между ними безусловно есть связь, так как прошлые соседи получены из текущих узлов, но не могу сказать на сколько это решение плохое или хорошее для MySQL. А вот для PostgreSQL, как минимум на данном этапе жизни расширения, могу - оно плохое. Причина заключается в выборе функции создания плана - `make_join_rel`.

Reviewer: ```suggestion
Как не трудно догадаться, после этого количество узлов (а значит и возможных пар csg/cmp), которые следует рассмотреть, увеличится, а значит и количество "мусорных" решений, которые либо не дадут полезного ответа, либо будут рассмотрены несколько раз, тоже увеличится. Между ними безусловно есть связь, так как прошлые соседи получены из текущих узлов, но не могу сказать на сколько это решение плохое или хорошее для MySQL. А вот для PostgreSQL, как минимум на данном этапе жизни расширения, могу — оно плохое. Причина заключается в выборе функции создания плана — `make_join_rel`.
```

> Вызов этой функции очень дорого стоит и занимает почти все время работы. Если использовать подход MySQL в этой части то по итогу будет создано довольно много ненужных пар CSG/CMP, для которых мы должны будем создать план. И мы, скорее всего, потратим свое время и ресурсы на это, потому что, даже если между отношениями нет предикатов, то мы можем создать `CROSS JOIN`. Короче говоря, в текущей модели стоимости, гораздо выгоднее потратиться на лишний вызов функции нахождения соседей (`get_neighbors`), чем на создание плана (`make_join_rel`).

Reviewer: ```suggestion
Вызов этой функции очень дорого стоит и занимает почти все время работы. Если использовать подход MySQL в этой части, то по итогу будет создано довольно много ненужных пар CSG/CMP, для которых мы должны будем создать план. И мы, скорее всего, потратим свое время и ресурсы на это, потому что даже если между отношениями нет предикатов, то мы можем создать `CROSS JOIN`. Короче говоря, в текущей модели стоимости, гораздо выгоднее потратиться на лишний вызов функции нахождения соседей (`get_neighbors`), чем на создание плана (`make_join_rel`).
```

Сказать, какой подход лучше нельзя:

1. В одних сценариях, имеет смысл использовать "идеальный" кеш (и адаптировать кеширование из MySQL):
   - Для большей части подмножеств у нас уже имеется созданный план
   - Дальнейших рекурсивных вызовов не предвидится (и не будет экспоненциального взрыва)
2. В других, таких планов вообще нет, поэтому соседей находить не надо вообще и можно изредка их честно вычислять

Но даже так, взяв первый подход, нам следует оценить последствия, так как добавление еще нескольких узлов в соседство увеличит дальнейшие расходы на другие итерации. Это может практически полностью нивелировать полученный прямо сейчас выигрыш - не забывайте, что каждый дополнительный узел увеличивает количество итераций по подмножествам *в 2 раза*, а этот дополнительный узел еще и добавит новых узлов в будущих соседей! Из примера выше - мы добавили 2 узла в соседи, а значит итераций станет в 4 раза больше и для каждого нужно будет найти еще соседей и вызвать другие функции и так далее. А сколько таких "косвенных" соседей накопится за еще большее количество рекурсивных вызовов? Вопрос риторический.

</spoiler>

## Индексирование ребер

> Во время рассказа о деталях реализации упомянул, что сложные ребра сортируются. Одна из причин - это избавление от дубликатов, но есть и другая. Чтобы понять какая, надо посмотреть на некоторые части DPhyp, а конкретно те места, в которых нам необходимо обходить ребра: нахождение соседей и определение связи (ребра) между 2 гиперузлами. Если приглядеться, то можно увидеть похожий паттерн с идеей кеширования - есть движущиеся детали, а есть постоянные:

Reviewer: ```suggestion
Во время рассказа о деталях реализации я упомянул, что сложные ребра сортируются. Одна из причин — это избавление от дубликатов, но есть и другая. Чтобы понять какая, надо посмотреть на некоторые части DPhyp, а конкретно на те места, в которых нам необходимо обходить ребра: нахождение соседей и определение связи (ребра) между 2 гиперузлами. Если приглядеться, то можно увидеть похожий паттерн с идеей кеширования — есть движущиеся детали, а есть постоянные:
```

- при поиске соседей множество исключенных узлов (`excluded`) фиксированно (либо только увеличивается)
- при определении связи между гиперузлами оба гиперузла фиксированы

Можно ли как-то использовать это знание? Почему бы и нет. Для начала научимся учитывать исключенные узлы.

> Проанализировав DPhyp, можно понять, что множество исключенных узлов имеет одинаковую структуру: лидирующие единицы, а затем разрозненные элементы, например, `010110011111` - есть 5 лидирующих единиц. Это не случайно, так работает алгоритм - мы не должны смотреть на узлы, которые еще не обрабатывали, поэтому каждый раз при обходе ребер мы проверяем, что никакая часть не пересекается с этими исключенными. Множество исключенных узлов в процессе итерирования может только увеличиться, но не уменьшится, а это значит, что мы заранее можем знать, что какие узлы точно не будут удовлетворять условию, то есть будут пересекаться с этими лидирующими нулями.

Reviewer: ```suggestion
Проанализировав DPhyp, можно понять, что множество исключенных узлов имеет одинаковую структуру: лидирующие единицы, а затем разрозненные элементы, например, `010110011111` — в нем есть 5 лидирующих единиц. Это не случайно, так работает алгоритм — мы не должны смотреть на узлы, которые еще не обрабатывали, поэтому каждый раз при обходе ребер мы проверяем, что никакая часть не пересекается с этими исключенными. Множество исключенных узлов в процессе итерирования может только увеличиться, но не уменьшиться, а это значит, что мы заранее можем знать, какие узлы точно не будут удовлетворять условию, то есть будут пересекаться с этими лидирующими нулями.
```

> Мы можем сделать следующее. Берем все гиперребра и сортируем их по правой части, а для сравнения используем количество лидирующих нулей. Затем, когда приходит время итерирования, мы вычисляем длину последовательности лидирующих единиц в множестве исключенных и стартовый индекс для итерации ставим таким, чтобы первый элемент правой части ребра был точно больше последнего элемента последовательности единиц.

Reviewer: ```suggestion
Мы можем сделать следующее. Берем все гиперребра и сортируем их по правой части, а для сравнения используем количество лидирующих нулей. Затем, когда приходит время итерирования, мы вычисляем длину последовательности лидирующих единиц во множестве исключенных и стартовый индекс для итерации ставим таким, чтобы первый элемент правой части ребра был точно больше последнего элемента последовательности единиц.
```

Если не понятно, покажу на примере. У нас имеется несколько отсортированных по количеству лидирующих нулей гиперребер. Справа написал индексы для удобства (левая часть гиперребра не важна, только правая):

```text
[
    xxxxx - 00101    0
    xxxxx - 00111    1
    xxxxx - 00100    2
    xxxxx - 00100    3
    xxxxx - 01100    4
    xxxxx - 10100    5
    xxxxx - 11100    6
    xxxxx - 01000    7
]
```

> Теперь мы с легкостью можем понять с какого индекса нам стоит начать итерацию в зависимости от того, какое множество исключенных узлов перед нами:

Reviewer: ```suggestion
Теперь мы с легкостью можем понять, с какого индекса нам стоит начать итерацию в зависимости от того, какое множество исключенных узлов перед нами:
```

| `excluded` | стартовый индекс |
| ---------- | ---------------- |
| `00100`    | 0                |
| `00001`    | 2                |
| `11101`    | 2                |
| `10011`    | 2                |
| `00111`    | 7                |
| `01111`    | 8                |

Сразу несколько моментов:

> 1. Мы можем быть уверен только в размере лидирующих единиц, остальная часть для нас черный ящик. Например, из-за этого в 3 примере итерацию начинаем со 2 индекса, хотя остальная часть видно, что полностью исключена.

Reviewer: Наверно это должен быть маркированный список.
```suggestion
1. Мы можем быть уверены только в размере лидирующих единиц, остальная часть для нас черный ящик. Например, из-за этого в 3 примере итерацию начинаем со 2 индекса, хотя видно, что остальная часть полностью исключена.
```

> 2. Для обобщения кода, в случае когда нет подходящих ребер мы можем возвращать длину массива (большое число)

Reviewer: ```suggestion
2. Для обобщения кода в случае когда нет подходящих ребер мы можем возвращать длину массива (большое число).
```

> Чтобы понять ее пользу, вспомните, что все ребра в графе двунаправленные, то есть для каждого гиперребра мы создаем 2 пары с обмененными левой и правой частями. Поэтому, может оказаться, что на какой-то узел ссылается очень большое количество таблиц (например, это таблица фактов, а размерностей у нас 100+), а сама эта таблица имеет самый большой индекс, то есть `excluded` будет включать почти все таблицы. В этом случае, нам придется безуспешно проходить по всем сложным ребрам, зная, что это не даст никакого результата.

Reviewer: ```suggestion
Чтобы понять ее пользу, вспомните, что все ребра в графе двунаправленные, то есть для каждого гиперребра мы создаем 2 пары с обмененными левой и правой частями. Поэтому может оказаться, что на какой-то узел ссылается очень большое количество таблиц (например, это таблица фактов, а размерностей у нас 100+), а сама эта таблица имеет самый большой индекс, то есть `excluded` будет включать почти все таблицы. В этом случае, нам придется безуспешно проходить по всем сложным ребрам, зная, что это не даст никакого результата.
```

Хорошо, а как мы будем получать стартовый индекс? Первое что приходит в голову, когда говорят "отсортированный" - бинарный поиск, но не спешите с выводами. Вспомним, что наше пространство "ключей" (это количество лидирующих нулей/единиц) дискретное. Начинается оно с 0 и может только увеличиваться (в моем случае есть предел - 64). А что подходит под описание такой структуры? Массив. Под индексом `i` будет хранится первый индекс массива гиперребер, первый элемент правой части (гиперребра) которого не меньше этого `i`.

> А что делать с пробелами? Например, в примере после `00111` сразу идет `00100`, без `00010`. Эти пробелы мы заполняем предыдущим значением, то есть если для какой-то длины `i` я не знаю с какого индекса мне лучше начать итерацию, то нужно взять предыдущий индекс, так как если мой первый элемент `i`, то я буду удовлетворять и всем последовательностям исключенных длиной меньше этого `i`. За базу мы возьмем индекс `0`, то есть если такой последовательности исключенных узлов нет, - ее значение будет `0`, так как в этом случае мы должны проитерироваться по всем гиперребрам.

Reviewer: ```suggestion
А что делать с пробелами? Например, в примере после `00111` сразу идет `00100`, без `00010`. Эти пробелы мы заполняем предыдущим значением, то есть если для какой-то длины `i` я не знаю, с какого индекса мне лучше начать итерацию, то нужно взять предыдущий индекс, так как если мой первый элемент `i`, то я буду удовлетворять и всем последовательностям исключенных длиной меньше этого `i`. За базу мы возьмем индекс `0`, то есть если такой последовательности исключенных узлов нет — ее значение будет `0`, так как в этом случае мы должны проитерироваться по всем гиперребрам.
```

> Затраты на использование такого индекса равны только расчету количества лидирующих единиц в множестве исключенных узлов. Но это можно оптимизировать простым битовым трюком - прибавляем `1` и получаем последовательность из `0` такой же длины, но оканчивающуюся на `1`, а затем рассчитываем позицию этой единицы. Получить это значение можно специальной инструкцией, например, `POPCNT`.

Reviewer: ```suggestion
Затраты на использование такого индекса равны только расчету количества лидирующих единиц во множестве исключенных узлов. Но это можно оптимизировать простым битовым трюком - прибавляем `1` и получаем последовательность из `0` такой же длины, но оканчивающуюся на `1`, а затем рассчитываем позицию этой единицы. Получить это значение можно специальной инструкцией, например, `POPCNT`.
```

Для ребер из примера, мы можем построить такой индекс (индекс массива ребер справа):

```text
[
    0: 0
    1: 2
    2: 2
    3: 7
    4: 8
    5: 8
]
```

Размер этого индекса зависит только от максимального количества лидирующих `0` во всем массиве ребер. Для примера видно, что я мог бы создать массив только из 4 элементов, а дальше всегда возвращать `8` (как размер массива).

Мы научились быстро отсекать ненужные ребра при поиске соседей. Но по ребрам мы ходим также для определения связанности 2 гиперузлов. Можно ли использовать этот индекс и здесь? Да, можно.

> 2 гиперузла соединены, когда есть гиперребро левая часть которого подмножество левого гиперузла, а правая часть - правого. Гиперузлы на входе никак не меняются - точно также как и множество исключенных. А теперь надо заметить, что правая часть гиперребра точно не будет подмножеством, если в этой части есть элементы, индекс которых меньше, чем наименьший в гиперузле справа. Пример: правая часть гиперребра `001010` никак не будет подмножеством гиперузла `001100` из-за того, что в ребре есть элемент с индексом 2. То есть семантика здесь практически та же, что и в случае исключенного множества - мы должны исключать все ребра, у которых есть элементы меньше наименьшего элемента правого гиперузла.

Reviewer: ```suggestion
2 гиперузла соединены, когда есть гиперребро, левая часть которого — подмножество левого гиперузла, а правая часть — правого. Гиперузлы на входе никак не меняются, точно также как и множество исключенных. А теперь надо заметить, что правая часть гиперребра точно не будет подмножеством, если в этой части есть элементы, индекс которых меньше, чем наименьший в гиперузле справа. Пример: правая часть гиперребра `001010` никак не будет подмножеством гиперузла `001100` из-за того, что в ребре есть элемент с индексом 2. То есть семантика здесь практически та же, что и в случае исключенного множества — мы должны исключать все ребра, у которых есть элементы меньше наименьшего элемента правого гиперузла.
```

Вычисление стартового индекса находится в функции `get_start_index`. Для удобства на вход она принимает не готовый индекс, а множество исключенных узлов. Эта функция также используется при определении связанности 2 гиперузлов - надо просто перед передачей аргумента из множества, представляющего правое гиперребро, вычесть 1, тогда последовательность нулей обратно станет последовательностью единиц, что по требуемой семантике одно и то же.

```c++
static int
get_start_index(EdgeArray *edges, bitmapword excluded)
{
    int index;
    int lowest_bit;

    lowest_bit = bmw_rightmost_one_pos(excluded + 1);

    if (edges->start_idx_size <= lowest_bit)
        return edges->size;

    index = (int)edges->start_idx[lowest_bit];
    return index;
}
```

## Определение сложности запроса

> PostgreSQL использует 2 алгоритма: DPsize и GEQO. Причем последний используется если в запросе таблиц больше, чем значение параметра `geqo_threshold`. Но почему именно количество таблиц? А потому что сложность DPsize определяется количеством таблиц - мы безусловно будем рассматривать все возможные комбинации. Но DPhyp другое дело, его сложность зависит от самого графа запроса. В оригинальной статье сравниваются производительности алгоритмов на некоторых типах запросов, но они не дают прямой ответ на то, как определять сложность запроса (а может я проглядел). Но зато этот ответ есть в другой статье - ["Adaptive Optimization of Large Join Queries"](https://db.in.tum.de/~radke/papers/hugejoins.pdf).

Reviewer: ```suggestion
PostgreSQL использует 2 алгоритма: DPsize и GEQO. Причем последний используется, если в запросе таблиц больше, чем значение параметра `geqo_threshold`. Но почему именно количество таблиц? А потому что сложность DPsize определяется количеством таблиц — мы безусловно будем рассматривать все возможные комбинации. Но DPhyp другое дело, его сложность зависит от самого графа запроса. В оригинальной статье сравниваются производительности алгоритмов на некоторых типах запросов, но они не дают прямой ответ на то, как определять сложность запроса (а может я проглядел). Но зато этот ответ есть в другой статье - ["Adaptive Optimization of Large Join Queries"](https://db.in.tum.de/~radke/papers/hugejoins.pdf).
```

> Эта статья предлагает мета-алгоритм, который комбинируя несколько разных алгоритмов JOIN'а позволяет создать планы запросов с количеством таблиц в несколько тысяч. Но сейчас не об этом, а о DPhyp. Авторы предлагают использовать его в простых запросах. Но что такое простой запрос? Например, если в запросе 100 таблиц, то это не значит, что DPhyp с ним не справится. Например, если граф запроса представляет простую цепочку, например, все предикаты в форме `Ti.x OP T(i + 1).x`, то план для него найти просто. Но вот если это клика (каждый с каждым), то даже 15 таблиц это уже много. Поэтому для DPhyp сложность надо определять не в количестве таблиц, в сложности графа запроса - **количестве связных подграфов**. Значение в 10000 связных подграфов как предел эффективности запроса. Это соответствует примерно 14 таблицам в клике.

Reviewer: Я всё ещё не знаю, что такое клик, надеюсь те, кто читают — знают.
```suggestion
Эта статья предлагает метаалгоритм, который, комбинируя несколько разных алгоритмов JOIN'а, позволяет создать планы запросов с количеством таблиц в несколько тысяч. Но сейчас не об этом, а о DPhyp. Авторы предлагают использовать его в простых запросах. Но что такое простой запрос? Например, если в запросе 100 таблиц, то это не значит, что DPhyp с ним не справится. Например, если граф запроса представляет простую цепочку, например, все предикаты в форме `Ti.x OP T(i + 1).x`, то план для него найти просто. Но вот если это клика (каждый с каждым), то даже 15 таблиц это уже много. Поэтому для DPhyp сложность надо определять не в количестве таблиц, а в сложности графа запроса - **количестве связных подграфов**. Значение в 10000 связных подграфов как предел эффективности запроса. Это соответствует примерно 14 таблицам в клике.
```

               ->  Materialize  (cost=0.00..3.69 rows=27 width=12) (actual time=0.000..0.001 rows=27 loops=81)
                     ->  Nested Loop  (cost=0.00..3.56 rows=27 width=12) (actual time=0.011..0.021 rows=27 loops=1)
                           ->  Nested Loop  (cost=0.00..2.18 rows=9 width=8) (actual time=0.007..0.011 rows=9 loops=1)
                                 ->  Seq Scan on t9  (cost=0.00..1.03 rows=3 width=4) (actual time=0.004..0.004 rows=3 loops=1)
                                 ->  Materialize  (cost=0.00..1.04 rows=3 width=4) (actual time=0.001..0.001 rows=3 loops=3)
                                       ->  Seq Scan on t10  (cost=0.00..1.03 rows=3 width=4) (actual time=0.003..0.003 rows=3 loops=1)
                           ->  Materialize  (cost=0.00..1.04 rows=3 width=4) (actual time=0.000..0.001 rows=3 loops=9)
                                 ->  Seq Scan on t8  (cost=0.00..1.03 rows=3 width=4) (actual time=0.003..0.004 rows=3 loops=1)
 Planning Time: 4.706 ms
 Execution Time: 1365.543 ms
(44 rows)
```

**4 миллисекунды**! При этом планы идентичные - прирост производительности почти в **600 раз**!

> Но это всего-лишь 1 небольшой запрос, нужно что-то  На чем тестировать гадать не будем - возьмем готовый JOB, Join Order Benchmarks. Он был представлен в работе ["How Good Are Query Optimizers, Really?"](https://vldb.org/pvldb/vol9/p204-leis.pdf) и использовался для сравнения планировщиков нескольких СУБД, включая сам PostgreSQL. Сам бенчмарк можно найти в [этом репозитории](https://github.com/gregrahn/join-order-benchmark).

Reviewer: ```suggestion
Но это всего-лишь 1 небольшой запрос, нужно что-то, на чем гадать не будем - возьмем готовый JOB, Join Order Benchmarks. Он был представлен в работе ["How Good Are Query Optimizers, Really?"](https://vldb.org/pvldb/vol9/p204-leis.pdf) и использовался для сравнения планировщиков нескольких СУБД, включая сам PostgreSQL. Сам бенчмарк можно найти в [этом репозитории](https://github.com/gregrahn/join-order-benchmark).
```

> > В самой работе производилось сравнение не скорости работы планировщика, а оценок, которые он дает. Запросы, которые там используются не сложные - это INNER equi-JOIN'ы (т.е. условия JOIN только равенство), хотя для оценки реальной производительности следует использовать самые различные нагрузки, включая всякие сложные `LEFT JOIN`, но пока этого достаточно.

Reviewer: ```suggestion
> В самой работе производилось сравнение не скорости работы планировщика, а оценок, которые он дает. Запросы, которые там используются, не сложные — это INNER equi-JOIN'ы (т.е. условия JOIN только равенство), хотя для оценки реальной производительности следует использовать самые различные нагрузки, включая всякие сложные `LEFT JOIN`, но пока этого достаточно.
```

                           ->  Materialize  (cost=0.00..1.04 rows=3 width=4)
                                 ->  Seq Scan on t10  (cost=0.00..1.03 rows=3 width=4)
 Planning Time: 3337.097 ms
(42 rows)
```

Время планирования действительно гораздо дольше, но внимательнее всмотритесь в план запроса. Первое, что надо заметить - стоимость меньше. За счет чего? А все из-за вот этой неприглядной строчки:

```text
                     ->  Nested Loop  (cost=0.00..2.18 rows=9 width=8)
                           ->  Seq Scan on t6  (cost=0.00..1.03 rows=3 width=4)
                           ->  Materialize  (cost=0.00..1.04 rows=3 width=4)
                                 ->  Seq Scan on t7  (cost=0.00..1.03 rows=3 width=4)
```

Да, DPsize смог *найти неявную связь* между отношениями, даже если они были по разную сторону операндов. DPhyp такого не сделает, так как это разные стороны ребра, а по его правилам такое делать запрещено - нельзя соединять отдельные узлы разных гиперузлов, если они по разные стороны гиперребра, только все вместе. Из этого можно заключить, что DPhyp - это очень хорошая, но все же *эвристика*. Но, к сожалению, и это не все проблемы.

> Для создания гиперребер используется 3 различных источника, но они не покрывают всех вариантов. Проблема кроется в `joinclauses`. Во время работы PostgreSQL создает все возможные варианты выражения с разным набором необходимых отношений с левой и правой части. Это позволяет рассмотреть разные варианты расположения выражения в дереве. Проблема в том, что индексы отношений, которые там используются могут относится не к таблицам, а к индексам узлов JOIN (`RangeTblEntry` типа `RTE_JOIN`). И они там не просто так - они дают "подсказки" планировщику о том, какие отношения можно использовать и как переопределить отношения.

Reviewer: ```suggestion
Для создания гиперребер используется 3 различных источника, но они не покрывают всех вариантов. Проблема кроется в `joinclauses`. Во время работы PostgreSQL создает все возможные варианты выражения с разным набором необходимых отношений с левой и правой части. Это позволяет рассмотреть разные варианты расположения выражения в дереве. Проблема в том, что индексы отношений, которые там используются, могут относится не к таблицам, а к индексам узлов JOIN (`RangeTblEntry` типа `RTE_JOIN`). И они там не просто так — они дают "подсказки" планировщику о том, какие отношения можно использовать и как переопределить отношения.
```

> Чтобы было понятно, что имеется ввиду посмотрим на такой запрос (взят из регресс тестов самого PostgreSQL):

Reviewer: ```suggestion
Чтобы было понятно, что имеется ввиду, посмотрим на такой запрос (взят из регресс тестов самого PostgreSQL):
```

select t1.* from
  text_tbl t1
  left join (select *, '***'::text as d1 from int8_tbl i8b1) b1
    left join int8_tbl i8
      left join (select *, null::int as d2 from int8_tbl i8b2) b2
      on (i8.q1 = b2.q1)
    on (b2.d2 = b1.q2)
  on (t1.f1 = b1.d1)
  left join int4_tbl i4
  on (i8.q2 = i4.f1);
```

Проблемным здесь является предикат `i8.q2 = i4.f1` - для него хранится *3* копии с разным набором отношений левой и правой части:

- `{3}       - {8}`
- `{3, 6}    - {8}`
- `{3, 6, 7} - {8}`

3 и 8 - это индексы, соответствующие таблицам `i8` и `i4`, соответственно. Но что это за 6 и 7? Это индексы JOIN'ов:

```sql
-- 6
(select *, '***'::text as d1 from int8_tbl i8b1) b1
    left join int8_tbl i8
      left join (select *, null::int as d2 from int8_tbl i8b2) b2
      on (i8.q1 = b2.q1)
    on (b2.d2 = b1.q2)

-- 7
text_tbl t1
  left join (select *, '***'::text as d1 from int8_tbl i8b1) b1
    left join int8_tbl i8
      left join (select *, null::int as d2 from int8_tbl i8b2) b2
      on (i8.q1 = b2.q1)
    on (b2.d2 = b1.q2)
  on (t1.f1 = b1.d1)
```

Это в действительности отражает ограничения: `i8.q2 = i4.f1` применяется не к самому `i8`, а к результату `LEFT JOIN`а.

## Итоги

> Как вы могли догадаться, расширение пока довольно сырое. Из-за того, что существующая инфраструктура заточена под особенности PostgreSQL для реализации одного функционала приходится искать обходные пути, а для другого работает скрипя зубами.

Reviewer: ```suggestion
Как вы могли догадаться, расширение пока довольно сырое. Из-за того, что существующая инфраструктура заточена под особенности PostgreSQL, для реализации одного функционала приходится искать обходные пути, а для другого работает скрипя зубами.
```

2. Несвязные подграфы требуют отдельной обработки - пользователь сам должен говорить расширению как действовать (параметр `cj_strategy`)
3. Оценка страдает из-за неоптимальных первых аргументов - порядок обрабатываемых пар отношений для JOIN'ов важен, но сейчас он не соотносится с тем, что делает встроенный планировщик
4. Довольно долгое выполнение - сейчас используется встроенный `make_join_rel`, но он занимает колоссальное количество времени

Но есть и хорошая новость - **все поправимо**. Эти ограничения продиктованы только лишь реализацией, то есть не фундаментальные ограничения. Никто не запрещает нам написать свой `make_join_rel`, оптимизированный под DPhyp. В крайнем случае, мы можем пропатчить ядро и добавить недостающую функциональность. Тем более, мы нашли, как минимум, 1 запрос, который смогли ускорить в сотни раз, что может открыть целую область применимости, а значит работа проделана не зря.

Всем слона.

Ссылки:

- [Расширение pg_dphyp](https://github.com/ashenBlade/pg_dphyp)
- [Dynamic Programming Strikes Back](https://15721.courses.cs.cmu.edu/spring2018/papers/16-optimizer2/p539-moerkotte.pdf) - DPhyp

> - [Adaptive Optimization of Very Large Join Queries](https://db.in.tum.de/~radke/papers/hugejoins.pdf) - планирование огромных запросов

Reviewer: Общие замечания, которые лучше поправить после обработки указанных выше:
- Вместо тире "—" везде используется дефис "-", лучше пройтись поиском " - " и заменить на " — ", где это необходимо.
- Все сокращения "т.к.", "т.п." и другие пишутся через пробел: "т. к.", "т. п.". Можно найти и заменить с помощью регулярки `(?<!\w)([а-я])\.([а-я])\.` → `$1. $2.` Или вообще написать полностью: кое-где у тебя написано "как так", "то есть" целиком.
- Ты пишешь то "кеш", то "кэш", выбери что-то одно и пиши одинаково. Правильным считается "кеш".
- "в множестве" писать неправильно, надо "во множестве". Пройдись по всему документу.
- В русском языке десятичные дроби разделяются запятой, а не точкой. Но PostgreSQL тебе выдаёт значения именно с точкой, так как он не русский. Я бы всё таки десятичные значения в обычном тексте писал через запятую. Вот регулярка, чтобы найти все такие значения `\b\d+\.\d+\b`. Там где `0.5Кб`, `1.5 секунд` и похожих — я бы заменил.
